---
output:
  bookdown::pdf_document2:
    keep_tex: yes
    toc: false
link-citations: true
fontsize: 10pt
---

```{r, include=FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  cache=FALSE,
  fig.pos="ht"
)
```

```{r fig-5, fig.cap="(ref:fig5)"}
knitr::include_graphics("figs/figure5.pdf", auto_pdf = TRUE)
```

(ref:fig5) **Milo identifies the compositional disorder in cirrhotic liver**
(A-B) UMAP embedding of 58358 cells from healthy (n = 5) and cirrhotic (n = 5) human livers. Cells are colored by cellular lineage (A) and injury condition (B)
(C) Graph representation of neighbourhoods identified by Milo. 
Nodes are neighbourhoods, coloured by their log fold change between cirrhotic and healthy samples. Non-DA neighbourhoods (FDR 10%) are coloured white, and sizes correspond to the number of cells in a neighbourhood. Graph edges depict the number of cells shared between adjacent neighbourhoods. The layout of nodes is determined by the position of the neighbourhood index cell in the UMAP embedding of single cells. 
(D) Beeswarm plot showing the distribution of log-fold change in abundance between conditions in neighbourhoods from different cell type clusters. DA neighbourhoods at FDR 10% are coloured. Cell types detected as DA through clustering by Ramachandran et al. (2019) are annotated in the left side bar.
(E) UMAP embedding and graph representation of neighbourhoods of 7995 cells from endothelial lineage.
(F) Heatmap showing average neighbourhood expression of genes differentially expressed between DA neighbourhoods in the endothelial lineage (572 genes). Expression values for each gene are scaled between 0 and 1. The top panel denotes the neighbourhood DA log fold-change.


