---
title: "Milo: ageing mouse thymus"
output: html_notebook
---

# Introduction

In our recent paper []() we found that age reduces the differentiation of TEC precursors towards the medulla lineage. This came from clustering and 
annotating our single-cell data, and finding that a specific cluster (Intertypical TEC) expand with age. Subsequently, we identifed a specific subset of 
Intertypical TEC within which this defect manifests, but only after performing a much larger experiment with ~90,000 single-cells; our original experiment 
contained just 2000 cells. I will use `Milo` to identify the differential abundance of these neighbourhoods, and the genes associated with these perturbed 
states.

```{r, echo=TRUE, warning=FALSE, message=FALSE}
### Set up a mock data set using simulated data
library(ggplot2)
library(igraph)
library(ggthemes)
library(biomaRt)
library(ggsci)
library(umap)
library(reshape2)
library(SingleCellExperiment)
library(scran)
library(scater)
library(igraph)
library(miloR)
library(cowplot)
library(limma)
library(RColorBrewer)
library(pheatmap)
library(irlba)
library(tidyr)
library(dplyr)
library(ggbeeswarm)
library(colorspace)
library(ComplexHeatmap)
library(circlize)
```


```{r, warning=FALSE, message=FALSE}
thymus.data <- readRDS("~/Dropbox/Milo/simulations/data/AgeThymus_Down1_RealDA_read.RDS")
thymus.meta <- thymus.data$meta
thymus.meta$AgeInt <- as.numeric(as.character(gsub(thymus.meta$Age, pattern="wk", replacement="")))
thymus.mylo <- thymus.data$mylo
```

I'm going to re-do the PCA on these data as I need the loadings to be able to project the neighbourhoods into the same space as the single cells. This means 
that the representation will be slightly different.

```{r, warning=FALSE, message=FALSE}
thymus.hvg <- read.table("~/Dropbox/AgeingExperiment/Thymus_HVG.tsv", sep="\t", header=TRUE, stringsAsFactors=FALSE)

thymus.pca <- prcomp_irlba(t(logcounts(thymus.mylo[rownames(thymus.mylo) %in% thymus.hvg$gene_id[thymus.hvg$HVG],])), 
                           n=50, center=TRUE, scale.=TRUE)
reducedDim(thymus.mylo, "PCA") <- thymus.pca$x
attr(reducedDim(thymus.mylo, "PCA"), "rotation") <- thymus.pca$rotation
```


```{r, warning=FALSE, message=FALSE, fig.height=4.15, fig.width=6.75}
sim1.umap <- umap(reducedDim(thymus.mylo, "PCA")[, c(1:15)],
                  n_components=2,
                  n_neighbors=30, metric='euclidean',
                  init='random', min_dist=0.25)

meta.df <- cbind(thymus.meta, sim1.umap$layout)
colnames(meta.df) <- c(colnames(thymus.meta), "UMAP1", "UMAP2")
# add the label annotation
meta.df$Cluster <- "Unknown"
meta.df$Cluster[meta.df$TFIDF.Cluster == 2] <- "Intertypical TEC"
meta.df$Cluster[meta.df$TFIDF.Cluster == 9] <- "Perinatal cTEC"
meta.df$Cluster[meta.df$TFIDF.Cluster == 3] <- "Mature cTEC"
meta.df$Cluster[meta.df$TFIDF.Cluster == 7] <- "Mature mTEC"
meta.df$Cluster[meta.df$TFIDF.Cluster == 1] <- "Post-Aire mTEC"
meta.df$Cluster[meta.df$TFIDF.Cluster == 5] <- "Tuft-like mTEC"
meta.df$Cluster[meta.df$TFIDF.Cluster == 6] <- "Proliferating TEC"
meta.df$Cluster[meta.df$TFIDF.Cluster == 8] <- "nTEC"
meta.df$Cluster[meta.df$TFIDF.Cluster == 10] <- "sTEC"

ggplot(meta.df, aes(x=UMAP1, y=UMAP2)) +
  geom_point(aes(colour=Cluster)) +
  theme_clean() +
  scale_colour_npg() +
  guides(colour=guide_legend(override.aes=list(size=3), title="TEC subtype"))
```

This is a UMAP representation of the ageing single-cell data, coloured by the original clustering. We can see the same plot across ages (our experimental 
variable of interest).

```{r, message=FALSE, warning=FALSE, fig.height=7.15, fig.width=9.75}
ggplot(meta.df, aes(x=UMAP1, y=UMAP2)) +
    geom_point(data=meta.df[, c("UMAP1", "UMAP2")],
               colour='grey80', alpha=0.5, size=0.5) +
    geom_point(aes(colour=Cluster)) +
    theme_clean() +
    scale_colour_npg() +
    facet_wrap(~AgeInt) +
    guides(colour=guide_legend(override.aes=list(size=3), title="TEC subtype"))

ggsave("~/Dropbox/Milo/figures/AgeingThymus_UMAP_Agefacet.png",
       height=7.15, width=9.75, dpi=300)
```

I'll now use `Milo` to build the graph, define neighbourhoods and perform DA testing.

```{r}
thymus.meta$ExpSamp <- paste(thymus.meta$Age, thymus.meta$SortDay, sep="_")
sim.reps <- unlist(lapply(strsplit(unique(thymus.meta$ExpSamp), split="_"), FUN=function(X) paste0(X[2]))) 
sim.cond <- unlist(lapply(strsplit(unique(thymus.meta$ExpSamp), split="_"), FUN=function(X) paste0(X[1])))
testing.meta <- data.frame("Condition"=sim.cond, "Replicate"=sim.reps)  
testing.meta$Sample <- paste(testing.meta$Condition, testing.meta$Replicate, sep="_")
testing.meta$AgeInt <- as.numeric(as.character(gsub(testing.meta$Condition, pattern="wk", replacement="")))
```


```{r, warning=FALSE, message=FALSE}
thymus.mylo <- buildGraph(thymus.mylo, k=30, d=30, seed=42)
thymus.mylo <- makeNhoods(thymus.mylo, prop=0.25, k=30, d=15, refined=TRUE, reduced_dims="PCA")
thymus.mylo <- countCells(thymus.mylo, meta.data=thymus.meta, samples="ExpSamp")
```

What is the distribution of neighbourhood sizes?

```{r, warning=FALSE, message=FALSE, fig.height=3.25, fig.width=4.95}
plotNhoodSizeHist(thymus.mylo)
```

This shows that the mean neighbourhood size is ~50-60, that should give us enough cell counts for DA testing.

```{r, warning=FALSE, message=FALSE}
thymus.res <- testNhoods(thymus.mylo, design=~AgeInt, design.df=testing.meta, fdr.weighting="k-distance")
thymus.res$Diff <- sign(thymus.res$logFC)
thymus.res$Diff[thymus.res$SpatialFDR > 0.01] <- 0
table(thymus.res$Diff)
```

From this we can see that 64 neighbourhoods are depleted with age and 77 are enriched - but what are they? We can project our neighbourhoods into the same 
space as our single-cells to visualise where these DA neighbourhoods are.

```{r, warning=FALSE, message=FALSE}
max.lfc <- max(abs(thymus.res$logFC))

ggplot(thymus.res, aes(x=logFC, y=-log10(SpatialFDR), colour=as.character(Diff))) +
    geom_hline(yintercept=-log10(0.01), lty=2, colour='grey50') +
    geom_point(size=3) +
    scale_x_continuous(limits=c(-max.lfc-0.1, max.lfc+0.1)) +
    theme_cowplot() +
    scale_colour_manual(values=c("firebrick2", "grey80", "firebrick2")) +
    guides(colour=FALSE) +
    labs(x="log Fold Change", y=expression(paste("-log"[10], " Spatial FDR")))

ggsave("~/Dropbox/Milo/figures/Thymus_Volcano.pdf",
       height=3.15, width=3.95, useDingbats=FALSE)
```

```{r, warning=FALSE, message=FALSE}
thymus.mylo <- calcNhoodExpression(thymus.mylo, subset.row=rownames(thymus.mylo) %in% thymus.hvg$gene_id[thymus.hvg$HVG])
thymus.mylo <- projectNhoodExpression(thymus.mylo, d=15, reduced_dims="PCA", scale=FALSE, center=TRUE)
```

First let's look at the UMAP of the projected neighbourhoods, then the joint UMAP with the single-cell data.

```{r, warning=FALSE, message=FALSE}
rownames(meta.df) <- meta.df$Sample
colData(thymus.mylo)$Cluster <- factor(meta.df[colnames(thymus.mylo), ]$Cluster)
colData(thymus.mylo)$Age <- ordered(meta.df[colnames(thymus.mylo), ]$AgeInt, levels=c(1, 4, 16, 32, 52))
colData(thymus.mylo)$SortDay <- meta.df[colnames(thymus.mylo), ]$SortDay

add_nhood_coldata_to_res <- function(thymus.mylo, milo_res, coldata_col){
  nhood_counts <- sapply(seq_along(nhoods(thymus.mylo)), function(x) table(colData(thymus.mylo)[as.vector(nhoods(thymus.mylo)[[x]]), coldata_col]))
  nhood_counts <- t(nhood_counts)
  rownames(nhood_counts) <- seq_along(nhoods(thymus.mylo))
  max_val <- apply(nhood_counts, 1, function(x) colnames(nhood_counts)[which.max(x)])
  max_frac <- apply(nhood_counts, 1, function(x) max(x)/sum(x))
  milo_res[coldata_col] <- max_val
  milo_res[paste0(coldata_col, "_fraction")] <- max_frac
  return(milo_res)
}

thymus.res <- add_nhood_coldata_to_res(thymus.mylo, thymus.res, "Age")
thymus.res <- add_nhood_coldata_to_res(thymus.mylo, thymus.res, "Cluster")
```


```{r, warning=FALSE, message=FALSE}
thymus.nhood.umap <- umap(nhoodReducedDim(thymus.mylo)[, c(1:15)],
                          n_components=2,
                          n_neighbors=30, metric='cosine',
                          init='random', min_dist=0.7)

is.nhood <- grepl(rownames(nhoodReducedDim(thymus.mylo)), pattern="nh")
thymus.nhood.meta <- cbind(thymus.res, thymus.nhood.umap$layout[is.nhood, ])
colnames(thymus.nhood.meta) <- c(colnames(thymus.res), "NH.UMAP1", "NH.UMAP2")
```


```{r, warning=FALSE, message=FALSE}
ggplot(thymus.nhood.meta, aes(x=NH.UMAP1, y=NH.UMAP2)) +
    geom_point(data=thymus.nhood.meta[thymus.nhood.meta$SpatialFDR > 0.01, ],
               size=1, colour='grey80', alpha=0.5) +
    geom_point(data=thymus.nhood.meta[thymus.nhood.meta$SpatialFDR <= 0.01, ],
               size=3, aes(colour=logFC)) +
    theme_clean() +
    scale_colour_gsea()
```

This shows us the logFC of the DA neighbourhoods - but which populations and states do these correspond to?

```{r, warning=FALSE, message=FALSE, fig.height=4.15, fig.width=6.75}
thymus.notNH.umap <- as.data.frame(thymus.nhood.umap$layout[!is.nhood, ])
colnames(thymus.notNH.umap) <- c("NH.UMAP1", "NH.UMAP2")
thymus.notNH.umap$Sample <- rownames(thymus.nhood.umap$layout[!is.nhood, ])
thymus.notNH.meta <- merge(meta.df, thymus.notNH.umap, by='Sample')
```

We can also visualise this by breaking down the neighbourhoods into their respective clusters.

```{r, warning=FALSE, message=FALSE, fig.height=4.15, fig.width=6.75}
ggplot(mapping=aes(x=NH.UMAP1, y=NH.UMAP2)) +
    geom_point(data=thymus.notNH.meta, shape=21,
               aes(fill=Cluster)) +
    theme_clean() +
    scale_fill_npg() +
    scale_colour_gsea() +
    guides(fill=guide_legend(override.aes=list(alpha=1, size=3))) +
    labs(x="UMAP1", y="UMAP2")

ggsave("~/Dropbox/Milo/figures/AgeingThymus_UMAP.png",
       height=4.15, width=6.75, dpi=300)
```


```{r, message=FALSE, warning=FALSE, fig.height=7.15, fig.width=9.75}
ggplot(mapping=aes(x=NH.UMAP1, y=NH.UMAP2)) +
    geom_point(data=thymus.notNH.meta[, c("NH.UMAP1", "NH.UMAP2")],
               colour='grey80', alpha=0.5, size=0.5) +
    geom_point(data=thymus.notNH.meta, shape=21,
               aes(fill=Cluster)) +
    theme_clean() +
    scale_fill_npg() +
    scale_colour_gsea() +
    facet_wrap(~AgeInt) +
    guides(fill=guide_legend(override.aes=list(alpha=1, size=3))) +
    labs(x="UMAP1", y="UMAP2")

ggsave("~/Dropbox/Milo/figures/AgeingThymus_UMAP_Agefacet.png",
       height=7.15, width=9.75, dpi=300)
```


```{r, warning=FALSE, message=FALSE, fig.height=4.15, fig.width=6.75}
ggplot(mapping=aes(x=NH.UMAP1, y=NH.UMAP2)) +
    geom_point(data=thymus.nhood.meta[thymus.nhood.meta$SpatialFDR > 0.01, ],
               size=1, colour='grey80', alpha=0.5) +
    geom_point(data=thymus.notNH.meta, shape=21, alpha=0.1,
               aes(fill=Cluster)) +
    geom_point(data=thymus.nhood.meta[thymus.nhood.meta$SpatialFDR <= 0.01, ],
               size=3, aes(colour=logFC)) +
    theme_clean() +
    scale_fill_npg() +
    scale_colour_gradient2(low=darken('blue'), mid='white', high=darken('#ff0000')) +
    guides(fill=FALSE) +
    labs(x="UMAP1", y="UMAP2")

ggsave("~/Dropbox/Milo/figures/AgeingThymus_NhoodUMAP_logFC.png",
       height=4.15, width=6.75, dpi=300)
```

These UMAPs are all well and good, but what about the exact clusters?

```{r, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=9.95}
ggplot(mapping=aes(x=Cluster, y=logFC)) +
    geom_quasirandom(data=thymus.res[thymus.res$SpatialFDR > 0.01,], 
                     alpha=0.75, colour='grey50', size=2) + 
    geom_quasirandom(data=thymus.res[thymus.res$SpatialFDR <= 0.01,], 
                     aes(colour=logFC), size=2) + 
    coord_flip() +
    facet_wrap(~Cluster, scales="free_y") +
    scale_colour_gradient2(low=darken('blue'), mid='grey80', high=darken('#ff0000')) +
    expand_limits(y=c(0)) +
    theme_clean()

ggsave("~/Dropbox/Milo/figures/Thymus_ResByCluster.pdf",
       height=4.95, width=9.95, useDingbats=FALSE)
```

That's super interesing! There is a subset of `Intertypical TEC` that are diminished with age - we definitely didn't see that until the lineage-tracing 
experiments. I wonder if these are the mTEC-biased precursors - the gene expression data will be useful here!

To perform DE testing across DA neighbourhoods I'll merge the adjacent neighbourhoods with the same logFC sign that are significantly DA. For the purposes 
of this illustration I'll use just the Intertypical TEC.

```{r, warning=FALSE, message=FALSE}
itec.mylo <- thymus.mylo[, colData(thymus.mylo)$Cluster %in% c("Intertypical TEC")]
itec.res <- thymus.res[thymus.res$Cluster %in% c("Intertypical TEC"), ]
```

I'll require adjacent neighbourhoods to share at least one cell.

```{r, warning=FALSE, message=FALSE}
.group_nhoods_by_overlap <- function(nhs){
  ll_names <- expand.grid(names(nhs), names(nhs)) 
  
  lintersect <- sapply(1:nrow(ll_names), function(x) length(intersect(nhs[[ll_names[x,1]]], nhs[[ll_names[x,2]]]))) 
  ## Count as connected only nhoods with at least n shared cells
  n=5
  lintersect_filt <- ifelse(lintersect < n, 0, lintersect)
  
  ## Convert to adjacency matrix (values = no of common cells)
  d <- matrix(lintersect_filt, nrow = length(nhs), byrow = TRUE)
  
  g <- graph_from_adjacency_matrix(d)
  groups <- components(g)$membership
  return(groups)  
}

.cells_in_nhood_groups <- function(nhs, gr){
  ls <- list()
  for (g in unique(gr)){
    nhs_ls <- lapply(nhs[which(gr==g)], function(x) as.vector(x))
    cells <- purrr::reduce(nhs_ls, union)
    ls[[g]] <- cells
  }
  return(ls)
}
```


```{r, warning=FALSE, message=FALSE}
pos_nhoods <- itec.res[itec.res$SpatialFDR <= 0.01 & itec.res$logFC > 0, ]$Nhood
neg_nhoods <- itec.res[itec.res$SpatialFDR <= 0.01 & itec.res$logFC < 0, ]$Nhood
```


```{r, warning=FALSE, message=FALSE}
group_negs <- .group_nhoods_by_overlap(nhoods(itec.mylo)[neg_nhoods])
cells_neg <- rownames(colData(itec.mylo))[unlist(.cells_in_nhood_groups(nhoods(itec.mylo)[neg_nhoods], group_negs))]
cells_neg <- cells_neg[!is.na(cells_neg)]
neg.df <- data.frame("CellID"=cells_neg, "nhood_group"=rep(paste0("negLogFC_", unique(group_negs)), length(unlist(cells_neg))))

group_pos <- .group_nhoods_by_overlap(nhoods(itec.mylo)[pos_nhoods])
cells_pos <- rownames(colData(itec.mylo))[unlist(.cells_in_nhood_groups(nhoods(itec.mylo)[pos_nhoods], group_pos))]
cells_pos <- cells_pos[!is.na(cells_pos)]
pos.df <- data.frame("CellID"=cells_pos, "nhood_group"=rep(paste0("posLogFC_", unique(group_pos)), length(unlist(cells_pos))))

test_groups_df <- do.call(rbind.data.frame, list("neg"=neg.df, "pos"=pos.df))
test_groups_df <- test_groups_df[!duplicated(test_groups_df$CellID), ]

colData(itec.mylo)$test_groups <- NA
colData(itec.mylo)[unique(test_groups_df$CellID), ]$test_groups <- as.character(test_groups_df$nhood_group)
```


```{r, warning=FALSE, message=FALSE}
# subset to just the HVGs
itec.hvg <- modelGeneVar(itec.mylo)
itec.hvgs <- getTopHVGs(itec.hvg, n=5000)

# find markers between logFC groups
itec.meta <- colData(itec.mylo)[!is.na(colData(itec.mylo)$test_groups), ]
itec.exprs <- logcounts(itec.mylo[itec.hvgs, !is.na(colData(itec.mylo)$test_groups)])

# itec.design <- model.matrix(object = ~ 0 + test_groups, data=itec.meta)
itec.design <- model.matrix(object = ~ Age, data=itec.meta)
#itec.constrasts <- makeContrasts(test_groupsposLogFC_1 - test_groupsnegLogFC_1, levels=itec.design)
itec.fit <- lmFit(itec.exprs, itec.design)
#itec.fit <- contrasts.fit(itec.fit, itec.constrasts)
itec.fit <- eBayes(itec.fit)

itec.markers <- as.data.frame(topTable(itec.fit, coef=2, number = Inf, sort.by = "p", p.value = 1))
itec.markers$adj.P.Val[is.na(itec.markers$adj.P.Val)] <- 1
itec.markers$logFC[is.infinite(itec.markers$logFC)] <- 0

itec.markers$Sig <- as.character(as.numeric(itec.markers$adj.P.Val <= 0.01))
itec.markers$Diff <- sign(itec.markers$logFC)
itec.markers$Diff[itec.markers$Sig == 0] <- 0
table(itec.markers$Diff)
```

These are the genes that are DE across ages in the Intertypical TEC and amongst the top 5000 HVGs at 1% FDR. For each DE gene I'll calculate the % of cells 
in the corresponding neighbourhood that express that gene.

```{r, warning=FALSE, message=FALSE}
.calculate_nhood_perc_expression <- function(milo, nhoods, gene){
    perc_expr <- sapply(nhoods(milo)[nhoods], FUN=function(X) sum(logcounts(milo[gene, as.numeric(X)]) > 0)/length(X))
    perc_expr <- setNames(perc_expr, nhoods)
    
    sink(file="/dev/null")
    gc()
    sink(file=NULL)
    
    return(perc_expr)
}
```


```{r, warning=FALSE, message=FALSE}
biomaRt.connection <- useMart("ensembl", "mmusculus_gene_ensembl")

gene.df <- getBM(attributes = c("ensembl_gene_id", "external_gene_name", "chromosome_name"),
                 filters="ensembl_gene_id", 
                 values=rownames(thymus.mylo), mart=biomaRt.connection)
rownames(gene.df) <- gene.df$ensembl_gene_id
```

```{r, warning=FALSE, message=FALSE}
thymus.mylo <- calcNhoodExpression(thymus.mylo, assay="logcounts")
```

```{r, warning=FALSE, message=FALSE}
goi.genes <- c("Cd52", "Ccl21a", "Psmb11", "Ly6a", "Aire", "Cd40")
goi.ensembl <- gene.df[gene.df$external_gene_name %in% goi.genes, ]$ensembl_gene_id
sig.markers <- itec.markers[itec.markers$adj.P.Val <= 0.01, ]
top.genes <- rownames(sig.markers[order(sig.markers$logFC, decreasing=TRUE), ])[1:25]
bottom.genes <- rownames(sig.markers[order(sig.markers$logFC, decreasing=FALSE), ])[1:25]
plot.genes <- unique(c(top.genes, bottom.genes, goi.ensembl))
gene.symbols <- gene.df[plot.genes, ]$external_gene_name
```

I think I might be able to cluster the heatmap rows to make this more intuitive.

```{r, warning=FALSE, message=FALSE, fig.height=8.95, fig.width=7.95}
hm.mat <- nhoodExpression(thymus.mylo)[plot.genes, ]
rownames(hm.mat) <- gene.df[plot.genes, ]$external_gene_name
colnames(hm.mat) <- c(1:ncol(hm.mat))
nhood.order <- itec.res$Nhood[order(itec.res$logFC, decreasing=FALSE)]
point.col <- rep("black", nrow(itec.res))
point.col[itec.res$SpatialFDR <= 0.01] <- "red"
point.col <- point.col[order(itec.res$logFC, decreasing=FALSE)]

hm.cols <- viridis::viridis(n=12, option="magma")

rank_annot <- HeatmapAnnotation(logFC=anno_points(itec.res[order(itec.res$logFC, decreasing=FALSE), ]$logFC,
                                                  gp=gpar(col=point.col)),
                                height=unit(2.5, "cm"))

gene_annot <- rowAnnotation(Genes=anno_mark(at=which(rownames(hm.mat) %in% c("Cd40", "Ccl21a", "Cd52", "Aire", "Ly6a")),
                                            labels=rownames(hm.mat)[rownames(hm.mat) %in% c("Cd40", "Ccl21a", "Cd52", "Aire", "Ly6a")]))

pdf("~/Dropbox/Milo/figures/AgeingThymus_iTEC_DEgenes-ComplexHeatmap.pdf",
    height=5.95, width=7.95, useDingbats=FALSE)
Heatmap(hm.mat[, nhood.order], cluster_columns=FALSE, show_column_names=FALSE, top_annotation=rank_annot,
        col=hm.cols, right_annotation=gene_annot, show_row_names = FALSE,
        row_title="Genes", column_title="Neighbourhoods",
        heatmap_legend_param = list(title="Expression"))
dev.off()

Heatmap(hm.mat[, nhood.order], cluster_columns=FALSE, show_column_names=FALSE, top_annotation=rank_annot,
        col=hm.cols, right_annotation=gene_annot, show_row_names = FALSE,
        row_title="Genes", column_title="Neighbourhoods",
        heatmap_legend_param = list(title="Expression"))
``` 


















