---
title: "Mouse thymus: Uncovering the decline of a fate-biased precursor state"
output: html_notebook
---

# Introduction

In our recent paper [Ageing compromises mouse thymus function and remodels epithelial cell differentiation](https://elifesciences.org/articles/56221) we found that age reduces the differentiation of TEC precursors 
towards the medulla lineage. This came from clustering and annotating our single-cell data, and finding that a specific cluster 
(Intertypical TEC) expand with age. Subsequently, we identifed a specific subset of Intertypical TEC within which this defect 
manifests, but only after performing a much larger experiment with ~90,000 single-cells; our original experiment contained just 
2000 cells. I will use `Milo` to identify the differential abundance of these neighbourhoods, and the genes associated with these 
perturbed states.

```{r, message=FALSE}
library(batchelor)
library(SingleCellExperiment)
library(miloR)
library(cowplot)
library(ggplot2)
library(igraph)
library(ggthemes)
library(biomaRt)
library(ggsci)
library(irlba)
library(umap)
library(viridis)
library(ComplexHeatmap)
library(ggbeeswarm)
library(colorspace)
library(patchwork)
library(enrichR)
library(scater)
library(ggsci)
```

We're loading in the data from an accompanying package that contains a few publicly available data sets. In this instance we've 
already packaged up the data into a `Milo` object. For an example of how to create a new `Milo` object, either from scratch or 
an existing `SingleCellExperiment` object, please see the vignette ...

```{r, warning=FALSE, message=FALSE}
thymus.data <- readRDS("~/Dropbox/Milo/simulations/data/AgeThymus_Down1_RealDA_read.RDS")
thymus.meta <- thymus.data$meta
thymus.meta$AgeInt <- as.numeric(as.character(gsub(thymus.meta$Age, pattern="wk", replacement="")))
thymus.mylo <- thymus.data$mylo
nhoodAdjacency(thymus.mylo) <- matrix(0L, ncol=1, nrow=1)
# remove ERCC genes

thymus.mylo <- thymus.mylo[!grepl(rownames(thymus.mylo), pattern="ERCC-"), ]
```

This shows the constituents of the `Milo` object - which is an extension of the `SingleCellExperiment` object. One of the things 
that we need to add to our object are the PCA rotations - this is used for projecting the neighbourhoods into the same PC-space 
as the single-cells (if we want to do that).

```{r, warning=FALSE, message=FALSE}
biomaRt.connection <- useMart("ensembl", "mmusculus_gene_ensembl")

gene.df <- getBM(attributes = c("ensembl_gene_id", "external_gene_name", "chromosome_name"),
                 filters="ensembl_gene_id", 
                 values=rownames(thymus.mylo), mart=biomaRt.connection)
rownames(gene.df) <- gene.df$ensembl_gene_id

rowData(thymus.mylo)$ensembl_gene_id <- rownames(thymus.mylo)
rowData(thymus.mylo)$symbol <- gene.df[rownames(thymus.mylo), ]$external_gene_name
rowData(thymus.mylo)$symbol[is.na(rowData(thymus.mylo)$symbol)] <- rowData(thymus.mylo)$ensembl_gene_id[is.na(rowData(thymus.mylo)$symbol)]
rownames(thymus.mylo) <- rowData(thymus.mylo)$symbol
```


```{r, warning=FALSE, message=FALSE}
thymus.hvg <- read.table("~/Dropbox/AgeingExperiment/Thymus_HVG.tsv", sep="\t", header=TRUE, stringsAsFactors=FALSE)

set.seed(42)
# the PCs are stored in the meta-data
# thymus.pca <- prcomp_irlba(t(logcounts(thymus.mylo[rownames(thymus.mylo) %in% thymus.hvg$gene_id[thymus.hvg$HVG],])), 
#                            n=50, center=TRUE, scale.=TRUE)
reducedDim(thymus.mylo, "PCA") <- as.matrix(thymus.meta[, paste0("PC", 1:50)])
```

We simply add the rotations as an attribute to the `reducedDim` slot in the `Milo` object.

The meta-data doesn't contain the labels we used in the original paper, so I'll add them here to make the interpretations clearer 
later.

```{r, warning=FALSE, message=FALSE, fig.height=4.15, fig.width=6.75}
# add the label annotation
thymus.meta$Cluster <- "Unknown"
thymus.meta$Cluster[thymus.meta$TFIDF.Cluster == 2] <- "Intertypical TEC"
thymus.meta$Cluster[thymus.meta$TFIDF.Cluster == 9] <- "Perinatal cTEC"
thymus.meta$Cluster[thymus.meta$TFIDF.Cluster == 3] <- "Mature cTEC"
thymus.meta$Cluster[thymus.meta$TFIDF.Cluster == 7] <- "Mature mTEC"
thymus.meta$Cluster[thymus.meta$TFIDF.Cluster == 1] <- "Post-Aire mTEC"
thymus.meta$Cluster[thymus.meta$TFIDF.Cluster == 5] <- "Tuft-like mTEC"
thymus.meta$Cluster[thymus.meta$TFIDF.Cluster == 6] <- "Proliferating TEC"
thymus.meta$Cluster[thymus.meta$TFIDF.Cluster == 8] <- "nTEC"
thymus.meta$Cluster[thymus.meta$TFIDF.Cluster == 10] <- "sTEC"

inter.cols <- c("#9970ab", "#35978f", "#B0cdc1", "#762a83", "#01665e", "#e7d4e8", "#dfc27d", "#8c510a" ,"#bf812d")
names(inter.cols) <- c("Post-Aire mTEC", 'Intertypical TEC','Mature cTEC', 'Tuft-like mTEC', 
                       'Proliferating TEC', 'Mature mTEC', 'nTEC', 'Perinatal cTEC', 'sTEC')
```

Now we are ready to begin the `Milo` workflow, which begins with building a kNN-graph. Prior to this point it is extremely 
important to remove any batch effects, and use a method or tool that returns either the corrected gene expression values, or a 
corrected reduced dimensional space. We require either the corrected gene expression or corrected reduced dimensions to compute 
the distances between cells for graph building and refining neighbourhoods.

For our mouse thymus data, we have the relevant PC co-ordinates already stored in the object. The first step is to construct 
the kNN-graph. For this data I'll use a small value for `k` as thymic epithelial cells are highly similar to each other. We can visualise this 
graph in a force-directed layout, in this case a Fruchterman-Reingold layout, implemented in `igraph`.

```{r, warning=FALSE, message=FALSE, fig.height=4.95, fig.width=6.15}
set.seed(42)
thymus.mylo <- buildGraph(thymus.mylo, k=21, d=40)

# what layout should I use? Force-directed?
fr.layout <- layout_with_fr(miloR::graph(thymus.mylo))
fr.df <- as.data.frame(fr.layout)
fr.df$Sample <- colnames(thymus.mylo)
fr.df <- merge(fr.df, thymus.meta, by='Sample')

# get the edges of the layout
fr.edges <- get.data.frame(miloR::graph(thymus.mylo))
fr.edges$from.Sample <- fr.edges$from
fr.edges$to.Sample <- fr.edges$to

# match edges to vertices and graph weights
fr.edges$from.x <- fr.df$V1[match(fr.edges$from.Sample, fr.df$Sample)]
fr.edges$from.y <- fr.df$V2[match(fr.edges$from.Sample, fr.df$Sample)]
fr.edges$to.x <- fr.df$V1[match(fr.edges$to.Sample, fr.df$Sample)]
fr.edges$to.y <- fr.df$V2[match(fr.edges$to.Sample, fr.df$Sample)]

ggplot(fr.df, 
       aes(x=V1, y=V2)) +
    geom_segment(data=fr.edges, aes(x=from.x, xend=to.x, y=from.y, yend=to.y),
                 colour='black') +
    geom_point(shape=21, size=4, aes(fill=Cluster)) +
    theme_cowplot() +
    scale_fill_manual(values=inter.cols) +
    theme(axis.title = element_blank(),
          axis.text=element_blank(),
          axis.ticks=element_blank(),
          axis.line=element_blank(),
          panel.grid=element_blank(),
          strip.text=element_text(size=12, colour='black'),
          legend.text=element_text(size=16), legend.title=element_text(size=18),
          legend.position="right",
          legend.direction="vertical") +
    guides(fill=guide_legend(title="TEC type", title.position="top")) +
    NULL

ggsave("~/Dropbox/Milo/figures/MouseThymus_FR_singleCells.png",
       height=4.95, width=6.15, dpi=300)
```

This is our layout, which shows the different separations of the TEC sub-types we described in our paper. We can also visualise this as a UMAP, which 
we can then use later on.

```{r}
set.seed(42)
sim1.umap <- umap(reducedDim(thymus.mylo, "PCA")[, c(1:40)],
                  n_components=2,
                  n_neighbors=21, metric='euclidean',
                  init='spectral', min_dist=0.3)

reducedDim(thymus.mylo, "UMAP") <- sim1.umap$layout # assign the UMAP co-ordinates to the reducedDim slot means we can access this later
meta.names <- colnames(thymus.meta)
thymus.meta <- cbind(thymus.meta, sim1.umap$layout)
colnames(thymus.meta) <- c(meta.names, "UMAP1", "UMAP2")
```


```{r, fig.height=4.95, fig.width=7.15}
by.type <- ggplot(thymus.meta, 
       aes(x=UMAP1, y=UMAP2)) +
    geom_point(size=3, aes(colour=Cluster)) +
    theme_cowplot() +
    scale_colour_manual(values=inter.cols) +
    theme(#axis.title = element_blank(),
          axis.text=element_blank(),
          axis.ticks=element_blank(),
          #axis.line=element_blank(),
          panel.grid=element_blank(),
          strip.text=element_text(size=12, colour='black'),
          legend.text=element_text(size=16), legend.title=element_text(size=18),
          legend.position="right",
          legend.direction="vertical") +
    guides(fill=guide_legend(title="TEC type", title.position="top")) +
    NULL

by.type
```

This is our UMAP with the cell type labels overlaid - we also generated the same plot but coloured by mouse age.

```{r, fig.height=4.95, fig.width=7.15}
age.cols <- viridis(option="magma", n=5)
names(age.cols) <- c("1wk", "4wk", "16wk", "32wk", "52wk")

by.age <- ggplot(thymus.meta, 
       aes(x=UMAP1, y=UMAP2)) +
    geom_point(size=3, aes(colour=Age)) +
    theme_cowplot() +
    scale_colour_manual(values=age.cols, labels=c("1wk", "4wk", "16wk", "32wk", "52wk")) +
    theme(#axis.title = element_blank(),
          axis.text=element_blank(),
          axis.ticks=element_blank(),
          #axis.line=element_blank(),
          panel.grid=element_blank(),
          strip.text=element_text(size=12, colour='black'),
          legend.text=element_text(size=16), legend.title=element_text(size=18),
          legend.position="right",
          legend.direction="vertical") +
    guides(fill=guide_legend(title="Age (weeks)",
                             title.position="top")) +
    NULL

by.age
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
# this generates plots that have aligned dimensions
aligned.plots <- align_patches(by.type, by.age)

ggsave(aligned.plots[[1]], filename="~/Dropbox/Milo/figures/MouseThymus_UMAP_singleCells.png",
       height=2.95, width=6.15, dpi=300)

ggsave(aligned.plots[[1]], filename="~/Dropbox/Milo/figures/MouseThymus_UMAP_singleCells.pdf",
       height=2.95, width=6.15, useDingbats=FALSE)


ggsave(aligned.plots[[2]], filename="~/Dropbox/Milo/figures/MouseThymus_UMAP_singleCells-Age.png",
       height=2.95, width=6.15, dpi=300)

ggsave(aligned.plots[[2]], filename="~/Dropbox/Milo/figures/MouseThymus_UMAP_singleCells-Age.pdf",
       height=2.95, width=6.15, useDingbats=FALSE)
```

Using our graph, we'll now assign cells to neighbourhoods and count the cells in each one according to our experimental design.

Our experiment was comprised of 5 time points: 1, 4, 16, 32 and 52 week old mouse thymi, with 5 replicate mice at each time point. This information 
is encapsulated in the meta-data column `ExpSamp`.

```{r}
thymus.meta$ExpSamp <- paste(thymus.meta$Age, thymus.meta$SortDay, sep="_")
table(thymus.meta$ExpSamp)
```


```{r, warning=FALSE, message=FALSE}
thymus.mylo <- makeNhoods(thymus.mylo, prop=0.3, k=21, d=40, refined=TRUE, reduced_dims="PCA")
thymus.mylo <- calcNhoodDistance(thymus.mylo, d=40)
```

We can plot the distribution of the neighbourhood sizes. Generally these want to fall in the range of 30-150 cells, but it depends on your 
experimental design if you need smaller or larger neighbourhoods. This can be changed by using a different value for $k$ when graph building and 
computing neighbourhoods.

```{r, warning=FALSE, message=FALSE, fig.height=3.25, fig.width=4.95}
plotNhoodSizeHist(thymus.mylo)
```

Most of our neighbourhoods contain about 30-40 cells; this should be sufficient for differential abundance testing with our experimental design.

```{r, warning=FALSE, message=FALSE}
colData(thymus.mylo)$ExpSamp <- thymus.meta[colnames(thymus.mylo), ]$ExpSamp
thymus.mylo <- countCells(thymus.mylo, meta.data=thymus.meta, samples="ExpSamp")
```

We now have a counts table stored in the `nhoodCounts` slot. You'll note that this is stored as sparse matrix - this is to help reduce the memory 
footprint for larger data sets.

```{r}
head(nhoodCounts(thymus.mylo))
```

For the differential abundance testing we need to construct a testing data frame which will tell `Milo` which samples come from which experimental 
variables, plus any covariates that you may wish to adjust for. In this case our data are fairly simple, so we are going to test the effect of age 
on neighbourhood abundance. Concretely, we will model the linear change in neighbourhood abundance with respect to age.

```{r}
sim.reps <- unlist(lapply(strsplit(unique(thymus.meta$ExpSamp), split="_"), FUN=function(X) paste0(X[2]))) 
sim.cond <- unlist(lapply(strsplit(unique(thymus.meta$ExpSamp), split="_"), FUN=function(X) paste0(X[1])))
testing.meta <- data.frame("Condition"=sim.cond, "Replicate"=sim.reps)  
testing.meta$Sample <- paste(testing.meta$Condition, testing.meta$Replicate, sep="_")
testing.meta$AgeInt <- as.numeric(as.character(gsub(testing.meta$Condition, pattern="wk", replacement="")))
```


```{r, warning=FALSE, message=FALSE}
rownames(testing.meta) <- testing.meta$Sample
thymus.res <- testNhoods(thymus.mylo, design=~AgeInt, design.df=testing.meta, fdr.weighting="k-distance")
thymus.res$Diff <- sign(thymus.res$logFC)
thymus.res$Diff[thymus.res$SpatialFDR > 0.1] <- 0
table(thymus.res$Diff)
```

We've used a fairly stringent FDR here, 1%, which gives us `r sum(thymus.res$Diff == -1)` neighbourhoods that are less abundant with age, and 
`r sum(thymus.res$Diff == 1)` that are more abundant with age. We can visualise these results in several different ways, such as a volcano plot to 
show the distribution of log fold changes with respect to the spatially adjusted P-values.

```{r, warning=FALSE, message=FALSE}
max.lfc <- max(abs(thymus.res$logFC))

ggplot(thymus.res, aes(x=logFC, y=-log10(SpatialFDR), colour=as.character(Diff))) +
    geom_hline(yintercept=-log10(0.1), lty=2, colour='grey50') +
    geom_point(size=3) +
    annotate("text", x=-0.175, y=2.5, label="Lower in older", size=4) +
    annotate("text", x=0.175, y=2.5, label="Higher in older", size=4) +
    scale_x_continuous(limits=c(-max.lfc-0.1, max.lfc+0.1)) +
    theme_cowplot() +
    scale_colour_manual(values=c("firebrick2", "grey80", "firebrick2")) +
    guides(colour=FALSE) +
    labs(x="log Fold Change", y=expression(paste("-log"[10], " Spatial FDR")))

ggsave("~/Dropbox/Milo/figures/MouseThymus_volcano.pdf",
       height=2.95, width=5.15, useDingbats=FALSE)
```

Another way to visualise this is to overlay these DA results on an abstracted graph or other reduced dimensional visualisation like a UMAP, that is 
anchored to the one we used for our single-cells above. There are a series of plotting functions baked into the `miloR` package that assist with this.

To do this we first build the neighbourhood graph using `buildNhoodGraph`, then pass our layout to the `plotNhoodGraph` function.

```{r, warning=FALSE, fig.height=5.95, fig.width=6.95}
rownames(fr.df) <- fr.df$Sample
reducedDim(thymus.mylo, "FR") <- fr.df[, c("V1", "V2")]
thymus.mylo <- buildNhoodGraph(thymus.mylo, overlap=5)
plotNhoodGraphDA(thymus.mylo, layout="UMAP", milo_res=thymus.res, alpha=0.1)

ggsave("~/Dropbox/Milo/figures/MouseThymus_UMAP_nhoods-logFC.png",
       height=6.95, width=7.55, dpi=300)

ggsave("~/Dropbox/Milo/figures/MouseThymus_UMAP_nhoods-logFC.pdf",
       height=6.95, width=7.55, useDingbats=FALSE)
```

This is like an abstracted graph where each node is a neighbourhood, the size of which corresponds to the number of cells in each and the edges are 
the numbers of cells shared between neighbourhoods. The nodes are coloured by the log fold change from the DA testing, where the non-DA 
neighbourhoods are filled in white. We can generate an equivalent plot where we colour by the original cluster identity for comparison.

This also demonstrates the flexibility of the plotting functions. The `plotNhoodGraph` functions return a `ggplot` object, which means they are 
customizable. In our case I'd like to use the same colour scheme we used in the paper, so I added a `scale_fill_manual` layer to the ggplot object. 
You can also tweak some of the other plotting aesthetics, such as making the legend points larger, etc.

```{r, warning=FALSE, fig.height=5.95, fig.width=6.95}
rownames(thymus.meta) <- thymus.meta$Sample
colData(thymus.mylo)$Cluster <- thymus.meta[colnames(thymus.mylo), ]$Cluster

plotNhoodGraph(thymus.mylo, layout="UMAP", colour_by="Cluster") +
    scale_fill_manual(values=inter.cols) +
    guides(fill=guide_legend(title="TEC type", override.aes=list(size=4)))

ggsave("~/Dropbox/Milo/figures/MouseThymus_UMAP_nhoods-cluster.png",
       height=5.95, width=6.95, dpi=300)

ggsave("~/Dropbox/Milo/figures/MouseThymus_UMAP_nhoods-cluster.pdf",
       height=5.95, width=6.95, useDingbats=FALSE)
```

Of course, this is not where the party ends. We also want to know exactly what these DA neighbourhoods represent. In reality, we rarely have nice 
annotated labels for our cells. To help facilitate the identification of cell states represented by neighbourhoods we can identify marker genes, 
analogous to how we'd do this for clusters. In our case we group together neighbourhoods with many overlapping cells and similar log fold-changes to 
try to get fairly similar groups. We can compare these directly to the original clusters too.

```{r, warning=FALSE, message=FALSE}
rownames(thymus.meta) <- thymus.meta$Sample
colData(thymus.mylo)$Cluster <- factor(thymus.meta[colnames(thymus.mylo), ]$Cluster)
colData(thymus.mylo)$Age <- ordered(thymus.meta[colnames(thymus.mylo), ]$AgeInt, levels=c(1, 4, 16, 32, 52))
colData(thymus.mylo)$SortDay <- thymus.meta[colnames(thymus.mylo), ]$SortDay

thymus.res <- annotateNhoods(thymus.mylo, thymus.res, "Age")
thymus.res <- annotateNhoods(thymus.mylo, thymus.res, "Cluster")
```


```{r, warning=FALSE, message=FALSE, fig.height=5.95, fig.width=4.15}
cluster.levels <- unique(thymus.res$Cluster)
thymus.res$Cluster <- factor(thymus.res$Cluster,
                             levels=cluster.levels)

change.levels <- c("Down", "Up", "Down", "Up", NA, NA, "Down", NA, NA)
names(change.levels) <- c("Mature mTEC", "Intertypical TEC", "Perinatal cTEC", "sTEC",
                          "Post-Aire mTEC", "Mature cTEC", "Proliferating TEC", "nTEC", "Tuft-like mTEC")

orig.da <- data.frame("Cluster"=cluster.levels,
                      "Direction"=change.levels[cluster.levels])
orig.da$start <- ifelse(orig.da$Direction == "Up", 0, 1)
orig.da$end <- ifelse(orig.da$Direction == "Down", 0, 1)

orig.da$Cluster <- factor(orig.da$Cluster,
                          levels=cluster.levels)

lim.val <- max(abs(thymus.res$logFC))
eps <- lim.val/25

thymus.bee <- ggplot(mapping=aes(x=Cluster, y=logFC)) +
    geom_quasirandom(data=thymus.res[thymus.res$SpatialFDR > 0.1,], 
                     alpha=1, colour='grey50', size=2) + 
    geom_quasirandom(data=thymus.res[thymus.res$SpatialFDR <= 0.1,], 
                     aes(colour=logFC), size=2.5) + 
    coord_flip() +
    facet_grid(Cluster~., scales="free_y", space="free") +
    scale_colour_gradient2(low=darken('#ff0000'), mid='grey80', high=darken('blue'),
                           midpoint=0) +
    scale_y_continuous(limits=c(-lim.val-eps, lim.val+eps)) +
    theme_bw() +
    labs(y="Log Fold Change") +
    theme(strip.text.y =  element_text(angle=0, colour='black', size=12),
          strip.background=element_rect(colour='white', fill='white'),
          axis.title.y = element_blank(), axis.text.y = element_blank(), 
          axis.text.x=element_text(size=12), axis.title.x=element_text(size=12),
          axis.ticks.y = element_blank()) +
    guides(colour=FALSE)


thymus.arrow <- ggplot(orig.da, aes(x=Cluster, y=start, xend=Cluster, yend=end, colour=Direction)) +
    geom_segment() +
    geom_segment(size=1.2, arrow=arrow(length=unit(0.1, "npc"), type="closed")) +
    coord_flip() +
    theme_bw() +
    scale_colour_manual(values=c(darken('blue'), darken('#ff0000')), breaks=c("Up", "Down"),
                        labels=c("Enriched with age", "Depleted with age")) +
    guides(colour=guide_legend(ncol=1, title="Baran-Gale et al.\nDA direction")) +
    facet_grid(Cluster~., scales="free_y", space="free") +
    labs(x="TEC type") +
    theme(strip.text.y = element_blank(), strip.text.x = element_text(angle=90),
        plot.margin = unit(c(0,0,0,0), "cm"), panel.grid = element_blank(),
        axis.title.x = element_blank(), axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(), axis.text.y=element_blank(),
        axis.title.y=element_text(size=12), 
        legend.text=element_text(size=12), legend.title=element_text(size=12),
        legend.position = "bottom")

(thymus.arrow + thymus.bee +
        plot_layout(widths=c(1, 8), guides="collect") & theme(legend.position="top", legend.justification=0)) +
    ggsave("~/Dropbox/Milo/revision/figures/Thymus_ResByCluster.pdf",
       height=5.95, width=4.15, useDingbats=FALSE)
```

To do this we need to pass in our `Milo` object and the results of our DA testing, plus any cell-wise or gene-wise subsetting we might wish to do. The 
`findNhoodMarkers` function will first aggregate neighbourhoods into groups based on the overlap provided and the same log fold change sign. Most 
of the interesting changes are happening in the `Intertypical TEC` cluster, so I'll subset down to those specific neighbourhoods. To do that we need 
to add a little annotation to each neighbourhood based on which cluster it most likely represents.

In practise, we identify marker genes for aggregated groups of neighbourhoods by fitting a linear model (either `limma` or `edgeR` depending on the 
assay used), which compares the average gene expression profile for the aggregated groups against all other neighbourhoods.

The `findNhoodMarkers` function returns `data.frame` containing paris of columns, one for each group of neighbourhoods, containing the log fold 
changes and adjusted P-values, along with a GeneID column. The format of the latter is the same as the input data, so in our case these are Ensembl 
gene IDs.

```{r}
thymus.mylo <- calcNhoodExpression(thymus.mylo) # This make plotting faster later.
```


```{r}
# keep.genes <- rownames(thymus.mylo)[rowMeans(logcounts(thymus.mylo[colData(thymus.mylo)$Cluster %in% c("Intertypical TEC")])) >= 0.1]
keep.genes <- rownames(thymus.mylo)[!grepl(rownames(thymus.mylo), pattern="(Gm)|(Rik)")]

# do the nhood grouping and add to DA results
thymus.res <- groupNhoods(thymus.mylo, thymus.res, da.fdr=0.1, overlap=5,
                          max.lfc.delta=0.1,
                          #subset.nhoods=thymus.res$Cluster %in% c("Intertypical TEC"),
                          compute.new=TRUE)
```

We can overlay the nhood groups onto the UMAP to see how they correspond to the clusters in the original manuscript.

```{r, fig.height=4.55, fig.width=9.95, warning=FALSE}
nhood.cols <- pal_d3()(9)
names(nhood.cols) <- c(1:9)

plotNhoodGroups(thymus.mylo, thymus.res) +
  guides(fill=guide_legend(ncol=4, override.aes=list(size=3))) +
  theme(legend.box="horizontal") +
  scale_fill_manual(values=nhood.cols) +
  guides(fill=guide_legend(title="Nhood Group", override.aes = list(size=3))) +
  ggsave("~/Dropbox/Milo/revision/figures/Mouse_Thymus_nhoodGroups.pdf",
         height=4.55, width=9.95, useDingbats=FALSE)
```


```{r}
table(thymus.res$Cluster, thymus.res$NhoodGroup)
```


This shows that the age-depleted Intertypical TEC are in neighbourhood group 5.

```{r}
thymus.markers <- findNhoodGroupMarkers(thymus.mylo, thymus.res,
                                        subset.nhoods=thymus.res$Cluster %in% c("Intertypical TEC"),
                                        subset.groups=c(3, 4, 5),
                                        subset.row=keep.genes, aggregate.samples = TRUE,
                                        sample_col="ExpSamp",
                                        gene.offset=FALSE) 
# thymus.markers$GeneID <- rownames(thymus.mylo):
# itec.groups <- thymus.markers$groups
# thymus.markers <- thymus.markers$dge
thymus.markers$GeneID <- keep.genes
head(thymus.markers)
```

We can visualise these results in a heatmap, which nicely illustrates the gradient of gene expression across neighbourhoods, ranked by their log 
fold changes. Due to the large number of genes, I'll subset to just those with a log fold change $\gt1$.

```{r, fig.height=5.95, fig.width=6.55, warning=FALSE}
plot.select <- c("Cd40", "Ccl21a", "Ly6a", "Stat1", "Il18r1", "Stat4", "Aff3", "Ptpn18")

select.markers <- unique(c(thymus.markers[thymus.markers$adj.P.Val_5 < 0.05 & thymus.markers$logFC_5 > 0, ]$GeneID, 
                           # thymus.markers[thymus.markers$adj.P.Val_2 < 0.01 & thymus.markers$logFC_2 > 1, ]$GeneID,
                           plot.select))

select.markers <- select.markers[!grepl(select.markers, pattern="(ENS)|(ERCC)")]
select.markers <- select.markers[select.markers != ""]

plotNhoodExpressionDA(thymus.mylo, da.res=thymus.res, alpha=0.1,
                      cluster_features=TRUE, scale_to_1=TRUE,
                      assay="logcounts", show_rownames=FALSE,
                      features=intersect(rownames(thymus.mylo), select.markers),
                      highlight_features=plot.select,
                      subset.nhoods=thymus.res$Cluster %in% c("Intertypical TEC")) +
  guides(fill=guide_colourbar(title="Scaled expression")) +
  theme(legend.title=element_text(size=12),
        legend.text=element_text(size=10),
        legend.key.size=unit(0.5, "cm"),
        legend.position=c(0, 0),
        # legend.margin=margin(0, 0, 0, 0),
        legend.box.margin=margin(0, 0, 0, 50))

ggsave("~/Dropbox/Milo/revision/figures/MouseThymus_iTEC_heatmap.pdf",
       height=5.95, width=6.55, useDingbats=FALSE)

ggsave("~/Dropbox/Milo/revision/figures/MouseThymus_iTEC_heatmap.png",
       height=5.95, width=6.55, dpi=90)
```

What are these DE genes between neighbourhoods? I'll perform a gene set enrichment analysis to see what pathways are enriched.

```{r}
dbs <- c("GO_Biological_Process_2018")
de.genes <- unique(c(thymus.markers[thymus.markers$adj.P.Val_5 < 0.1 & thymus.markers$logFC_5 > 0, ]$GeneID))
                     #thymus.markers[thymus.markers$adj.P.Val_2 < 0.01, ]$GeneID)) 
print(length(de.genes))
itec.enrichr <- enrichr(de.genes, dbs)$GO_Biological_Process_2018
head(itec.enrichr[itec.enrichr$Adjusted.P.value < 0.1, ])
```


