% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\hypersetup{
  pdftitle={Supplementary information for Differential cell-state abundance testing using k-NN graphs with Milo},
  pdfauthor={Emma Dann,; Neil C. Henderson,; Sarah A. Teichmann,; Michael D. Morgan,; John C. Marioni},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}
\urlstyle{same} % disable monospaced font for URLs
\usepackage[margin=1in]{geometry}
\usepackage{longtable,booktabs}
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
\newlength{\cslhangindent}
\setlength{\cslhangindent}{1.5em}
\newenvironment{cslreferences}%
  {}%
  {\par}

\title{Supplementary information for
\textbf{Differential cell-state abundance testing using k-NN graphs with \emph{Milo}}}
\author{Emma Dann, \and Neil C. Henderson, \and Sarah A. Teichmann, \and Michael D. Morgan, \and John C. Marioni}
\date{20 November, 2020}

\begin{document}
\maketitle

{
\setcounter{tocdepth}{3}
\tableofcontents
}
\renewcommand{\figurename}{Supplementary Figure}

\newpage

\hypertarget{supplementary-notes}{%
\section{Supplementary notes}\label{supplementary-notes}}

\hypertarget{description-of-milo}{%
\subsection{\texorpdfstring{Description of \emph{Milo}}{Description of Milo}}\label{description-of-milo}}

\hypertarget{building-the-k-nn-graph}{%
\subsubsection{Building the k-NN graph}\label{building-the-k-nn-graph}}

Similarly to many other tasks in single-cell analysis, \emph{Milo} uses a k-NN graph computed based on similarities in gene expression space as a representation of the phenotypic manifold in which cells lie. While \emph{Milo} can be used on graphs built with different similarity kernels, here we compute the graph as follows: a gene expression matrix of \(N\) cells is projected onto the first \(d\) principal components (PCs) to obtain a \(N \times d\) matrix \(X_{PC}\). Then, for each cell \(j\), the Euclidean distances to its \(k\) nearest neighbors in \(X_{PC}\) are computed and stored in a \(N \times N\) adjacency matrix. Then, \(D\) is made symmetrical, such that cells \(j\) and \(l\) are nearest neighbors (i.e.~connected by an edge) if either \(j\) is a nearest neighbor of \(l\) or \(l\) is a nearest neighbor of \(j\). The k-NN graph is encoded by the undirected symmetric version of \(\tilde{D}\) of \(D\), where each cell has at least K nearest neighbors.

\hypertarget{definition-of-cell-neighbourhoods-and-index-sampling-algorithm}{%
\subsubsection{Definition of cell neighbourhoods and index sampling algorithm}\label{definition-of-cell-neighbourhoods-and-index-sampling-algorithm}}

We define the neighbourhood \(n_i\) of cell \(j\) as the group of cells that are connected to \(j\) by an edge in the graph. Here, \(i=1,2,...,M\) indexes the sampled neighbourhoods, while \(j=1,2,...,N\) indexes the single-cells in the graph, so that \(M \leq N\).
Formally, a cell \(l\) belongs to neighbourhood \(n_i\) if \(\tilde{D}_{j,l} > 0\). We refer to \(j\) as the index cell of the neighbourhood.

In order to define a representative subset of neighbourhoods that span the whole k-NN graph, we implement a previously adopted algorithm to sample the index cells in a graph {[}\protect\hyperlink{ref-gutTrajectoriesCellcycleProgression2015}{1},\protect\hyperlink{ref-settyWishboneIdentifiesBifurcating2016}{2}{]}.
Briefly, we start by randomly sampling \(p \cdot N\) cells from the dataset, where \(p \in [0,1]\) (we use \(p = 0.1\) by default).
Given the reduced dimension matrix used for graph construction \(X_{PC}\), for each sampled cell we consider its \(k\) nearest neighbors \(l = 1,2,...,k\) with PC profiles \({x_1, x_2, ... , x_k}\). We measure the mean PC profile \(\bar{x}\) for the \(k\) cells and search for the cell \(j\) such that the Euclidean distance between \(x_j\) and \(\bar{x}\) is minimized. This yields a set of \(M \leq p \cdot N\) index cells that are used to define neighbourhoods.

\hypertarget{testing-for-differential-abundance-in-neighbourhoods}{%
\subsubsection{Testing for differential abundance in neighbourhoods}\label{testing-for-differential-abundance-in-neighbourhoods}}

\emph{Milo} builds upon the framework for differential abundance testing implemented by \emph{Cydar} {[}\protect\hyperlink{ref-lunTestingDifferentialAbundance2017}{3}{]}. In this section, we briefly describe the statistical model and adaptations to the k-NN graph setting.

\hypertarget{quasi-likelihood-negative-binomial-generalized-linear-models}{%
\paragraph*{Quasi-likelihood negative binomial generalized linear models}\label{quasi-likelihood-negative-binomial-generalized-linear-models}}
\addcontentsline{toc}{paragraph}{Quasi-likelihood negative binomial generalized linear models}

We consider a neighbourhood \(n\) with cell counts \(y_{ns}\) for each experimental sample \(s\). The counts are modelled by the negative binomial (NB) distribution, as it is supported over all non-negative integers and can accurately model both small and large cell counts. For such non-Normally distributed data we use generalized-linear models (GLMs) as an extension of classic linear models that can accomodate complex experimental designs. We therefore assume that
\[
y_{ns} \sim NB(\mu_{ns},\phi_{n}),
\]
where \(\mu_{ns}\) is the mean and \(\phi_{n}\) is the NB dispersion parameter.
The expected count value for neighbourhood \(n\) in experimental sample \(s\), \(\mu_{ns}\) is given by
\[
\mu_{ns} = \lambda_{ns}N_s
\]

where \(\lambda_{ns}\) is the proportion of cells belonging to experimental sample \(s\) in \(n\) and \(N_s\) is the total number of cells of \(s\). In practice, \(\lambda_{ns}\) represents the biological variability that can be affected by treatment condition, age or any biological covariate of interest. We use a log-linear model to represent the influence of the biological condition on the expected counts in neighbourhoods:
\[
log\ \mu_{ns} = \sum_{g=1}^{G}x_{sg}\beta_{ng} + log\ N_s
\]
where \(x_{sg}\) is the covariate vector indicating the condition applied to sample \(s\) and \(\beta_{ng}\) is the regression coefficient by which the covariate effects are mediated for neighbourhood \(n\).

Estimation of \(\beta_{ng}\) for each \(n\) and \(g\) is performed by fitting the GLM to the count data for each neighbourhood, i.e.~by estimating the dispersion \(\phi_{n}\) that models the variability of cell counts for replicate samples for each neighbourhood. Dispersion estimation is performed using the quasi-likelihood method in \texttt{edgeR}{[}\protect\hyperlink{ref-robinsonEdgeRBioconductorPackage2010a}{4}{]}, where the dispersion is modelled from the GLM deviance and thereby stabilized with empirical Bayes shrinkage, to stabilize the estimates in the presence of limited replication.

\hypertarget{adaptation-of-spatial-fdr-to-neighbourhoods}{%
\paragraph*{Adaptation of Spatial FDR to neighbourhoods}\label{adaptation-of-spatial-fdr-to-neighbourhoods}}
\addcontentsline{toc}{paragraph}{Adaptation of Spatial FDR to neighbourhoods}

To control for multiple testing, we adapt the Spatial FDR method introduced by \(Cydar\) {[}\protect\hyperlink{ref-lunTestingDifferentialAbundance2017}{3}{]}. The Spatial FDR can be interpreted as the proportion of the union of neighbourhoods that is occupied by false-positive neighbourhoods. This accounts for the fact that some neighbourhoods are more densely connected than others. To control spatial FDR in the k-NN graph, we apply a weighted version of the Benjamini-Hochberg (BH) method. Briefly, to control for FDR at some threshold \(\alpha\) we reject null hypothesis \(i\) where the associated p-value is less than the threshold
\[
\max_i{p_{(i)}: p_{(i)}\le \alpha\frac{\sum_{l=1}^{i}w_{(l)}}{\sum_{l=1}^{n}w_{(l)}}}
\]
Where the weight \(w_{(i)}\) is the reciprocal of the neighbourhood connectivity \(c_i\). As a measure of neighbourhood connectivity, we use the Euclidean distance to the kth nearest neighbour of the index cell for each neighbourhood.

\hypertarget{accounting-for-batch-effects}{%
\subsection{Accounting for batch effects}\label{accounting-for-batch-effects}}

Comparing biological conditions often requires acquiring single-cell data from multiple samples, sometimes generated with different experimental conditions or protocols. This commonly introduces batch effects, which can have a substantial impact on the data composition and subsequently the topology of any k-NN graph computed across the single-cell data. Consequently, this will have an impact on the ability of Milo to resolve genuinely DA subpopulations between experimental conditions. To minimize the emergence of false positive and false negative results that might be introduced by such technical effects, we recommend the following procedure: (1) Batch effects should be mitigated with computational data integration. Defining the best tool for this task is beyond the scope of this work (a large number of integration methods have been reviewed and benchmarked in {[}\protect\hyperlink{ref-lueckenBenchmarkingAtlaslevelData2020}{5}--\protect\hyperlink{ref-tranBenchmarkBatcheffectCorrection2020}{7}{]}). However, users should consider the type of output obtained by their integration method of choice, which can be a corrected feature space, a joint embedding or an integrated graph. Using a methods that produces a graph (e.g.~BBKNN {[}\protect\hyperlink{ref-polanskiBBKNNFastBatch}{8}{]}, Conos {[}\protect\hyperlink{ref-barkasJointAnalysisHeterogeneous2019}{9}{]}) will lead to suboptimal results in DA testing with Milo as the refined neighbourhood search procedure will still be affected by the batch effect, as this relies on finding neighbors in PCA space. (2) Known technical or nuisance covariates should be introduced in the experimental design of DA testing, exploiting the flexible GLM framework used by Milo.

To demonstrate the use of including both steps in DA testing analysis, we simulated data for a continuous trajectory with two technical batches (Suppl.Fig.1A) and simulated a subpopulation of differential abundance between two conditions (Suppl.Fig.1B). We tested for DA between conditions on the k-NN graph generated before and after batch correction using the Mutual Nearest Neighbors (MNN) method {[}\protect\hyperlink{ref-haghverdiBatchEffectsSinglecell2018}{10}{]}. In both the corrected and uncorrected data, we examined the effect of accounting for the batch in the experimental design of the GLM (\texttt{design\ =\ \textasciitilde{}\ batch\ +\ condition}). We find that, while alone MNN correction increases the true positive rate, accounting for the nuisance covariate in the testing design significantly increases accuracy and power both in the uncorrected and corrected graph (Suppl. Fig. 1C-D). The best DA prediction performance was achieved combining batch correction and an explicit experimental design adjusting for such nuisance covariates (Suppl. Fig. 1D).

To demonstrate how these results apply to a real dataset with a complex experimental design, we used a single-cell atlas of mouse gastrulation {[}\protect\hyperlink{ref-pijuan-salaSinglecellMolecularMap2019}{11}{]}. These data consist of 116312 cells collected from 411 embryos in 36 samples, across 10 developmental stages from E6.5 to E8.5 (Suppl. Fig. 2B). The samples were sequenced in 3 batches (Suppl. Fig. 2A). We used Milo to test for DA across developmental time points, on the k-NN graph built before and after batch correction with MNN. We find that including the sequencing batch in the DA testing design matrix removes false positives in the uncorrected graph, while increasing the number of detected DA neighbourhoods in the MNN corrected graph (Suppl. Fig.2C).

Taken together these results illustrate that while Milo can be used to robustly account for batch effects during DA testing, optimal results are achieved after the removal of major batch effects that are not readily accounted for by inclusion as a covariate in a linear model.

\hypertarget{references}{%
\section*{References}\label{references}}
\addcontentsline{toc}{section}{References}

\hypertarget{refs}{}
\begin{cslreferences}
\leavevmode\hypertarget{ref-gutTrajectoriesCellcycleProgression2015}{}%
1. Gut, G., Tadmor, M.D., Pe'er, D., Pelkmans, L., and Liberali, P. (2015). Trajectories of cell-cycle progression from fixed cell populations. Nature Methods \emph{12}, 951--954.

\leavevmode\hypertarget{ref-settyWishboneIdentifiesBifurcating2016}{}%
2. Setty, M., Tadmor, M.D., Reich-Zeliger, S., Angel, O., Salame, T.M., Kathail, P., Choi, K., Bendall, S., Friedman, N., and Pe'er, D. (2016). Wishbone identifies bifurcating developmental trajectories from single-cell data. Nature Biotechnology \emph{34}, 637--645.

\leavevmode\hypertarget{ref-lunTestingDifferentialAbundance2017}{}%
3. Lun, A.T.L., Richard, A.C., and Marioni, J.C. (2017). Testing for differential abundance in mass cytometry data. Nature Methods \emph{14}, 707--709.

\leavevmode\hypertarget{ref-robinsonEdgeRBioconductorPackage2010a}{}%
4. Robinson, M.D., McCarthy, D.J., and Smyth, G.K. (2010). edgeR: A Bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics \emph{26}, 139--140.

\leavevmode\hypertarget{ref-lueckenBenchmarkingAtlaslevelData2020}{}%
5. Luecken, M.D., Büttner, M., Chaichoompu, K., Danese, A., Interlandi, M., Mueller, M.F., Strobl, D.C., Zappia, L., Dugas, M., and Colomé-Tatché, M. \emph{et al.} (2020). Benchmarking atlas-level data integration in single-cell genomics. bioRxiv, 2020.05.22.111161.

\leavevmode\hypertarget{ref-chazarra-gilFlexibleComparisonBatch2020}{}%
6. Chazarra-Gil, R., Dongen, S. van, Kiselev, V.Y., and Hemberg, M. (2020). Flexible comparison of batch correction methods for single-cell RNA-seq using BatchBench. bioRxiv, 2020.05.22.111211.

\leavevmode\hypertarget{ref-tranBenchmarkBatcheffectCorrection2020}{}%
7. Tran, H.T.N., Ang, K.S., Chevrier, M., Zhang, X., Lee, N.Y.S., Goh, M., and Chen, J. (2020). A benchmark of batch-effect correction methods for single-cell RNA sequencing data. Genome Biology \emph{21}, 12.

\leavevmode\hypertarget{ref-polanskiBBKNNFastBatch}{}%
8. Polański, K., Young, M.D., Miao, Z., Meyer, K.B., Teichmann, S.A., and Park, J.-E. BBKNN: Fast batch alignment of single cell transcriptomes. Bioinformatics.

\leavevmode\hypertarget{ref-barkasJointAnalysisHeterogeneous2019}{}%
9. Barkas, N., Petukhov, V., Nikolaeva, D., Lozinsky, Y., Demharter, S., Khodosevich, K., and Kharchenko, P.V. (2019). Joint analysis of heterogeneous single-cell RNA-seq dataset collections. Nat Methods \emph{16}, 695--698.

\leavevmode\hypertarget{ref-haghverdiBatchEffectsSinglecell2018}{}%
10. Haghverdi, L., Lun, A.T.L., Morgan, M.D., and Marioni, J.C. (2018). Batch effects in single-cell RNA-sequencing data are corrected by matching mutual nearest neighbors. Nature Biotechnology \emph{36}, 421--427.

\leavevmode\hypertarget{ref-pijuan-salaSinglecellMolecularMap2019}{}%
11. Pijuan-Sala, B., Griffiths, J.A., Guibentif, C., Hiscock, T.W., Jawaid, W., Calero-Nieto, F.J., Mulas, C., Ibarra-Soria, X., Tyser, R.C.V., and Ho, D.L.L. \emph{et al.} (2019). A single-cell molecular map of mouse gastrulation and early organogenesis. Nature \emph{566}, 490--495.
\end{cslreferences}

\end{document}
