% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{amsmath,amssymb}
\usepackage{lmodern}
\usepackage{ifxetex,ifluatex}
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\hypersetup{
  pdftitle={Supplementary information for Differential cell-state abundance testing using KNN graphs with Milo},
  pdfauthor={Emma Dann,; Neil C. Henderson,; Sarah A. Teichmann,; Michael D. Morgan,; John C. Marioni},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}
\urlstyle{same} % disable monospaced font for URLs
\usepackage[margin=1in]{geometry}
\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
\usepackage{lineno}
\linenumbers
\renewcommand{\baselinestretch}{1.8}
\ifluatex
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\newlength{\cslhangindent}
\setlength{\cslhangindent}{1.5em}
\newlength{\csllabelwidth}
\setlength{\csllabelwidth}{3em}
\newenvironment{CSLReferences}[2] % #1 hanging-ident, #2 entry spacing
 {% don't indent paragraphs
  \setlength{\parindent}{0pt}
  % turn on hanging indent if param 1 is 1
  \ifodd #1 \everypar{\setlength{\hangindent}{\cslhangindent}}\ignorespaces\fi
  % set entry spacing
  \ifnum #2 > 0
  \setlength{\parskip}{#2\baselineskip}
  \fi
 }%
 {}
\usepackage{calc}
\newcommand{\CSLBlock}[1]{#1\hfill\break}
\newcommand{\CSLLeftMargin}[1]{\parbox[t]{\csllabelwidth}{#1}}
\newcommand{\CSLRightInline}[1]{\parbox[t]{\linewidth - \csllabelwidth}{#1}\break}
\newcommand{\CSLIndent}[1]{\hspace{\cslhangindent}#1}

\title{Supplementary information for
\textbf{Differential cell-state abundance testing using KNN graphs with \emph{Milo}}}
\author{Emma Dann, \and Neil C. Henderson, \and Sarah A. Teichmann, \and Michael D. Morgan, \and John C. Marioni}
\date{29 March, 2021}

\begin{document}
\maketitle

{
\setcounter{tocdepth}{3}
\tableofcontents
}
\renewcommand{\figurename}{Supplementary Figure}

\newpage

\hypertarget{supplementary-notes}{%
\section{Supplementary notes}\label{supplementary-notes}}

\hypertarget{description-of-workflow-for-milo-analysis}{%
\subsection{\texorpdfstring{Description of workflow for \emph{Milo} analysis}{Description of workflow for Milo analysis}}\label{description-of-workflow-for-milo-analysis}}

Given a single-cell dataset of gene expression profiles of \(L\) cells collected from \(S\) experimental samples, \emph{Milo} aims to quantify systematic changes in the abundance of cells between biological conditions. Here we provide a step-by-step description of the workflow for differential abundance analysis. Of note, we focus on the application to single-cell gene expression profiles, and we provide guidelines for pre-processing on this type of data. However, the core of the \emph{Milo} framework, from KNN graph construction to differential abundance testing, is applicable to any kind of single-cell dataset that can be embedded in a low-dimensional space.

\hypertarget{pp}{%
\subsubsection{Preprocessing and dimensionality reduction}\label{pp}}

For pre-processing of scRNA-seq profiles we recommend following standard practices in single-cell analysis {[}\protect\hyperlink{ref-lueckenCurrentBestPractices2019}{1},\protect\hyperlink{ref-amezquitaOrchestratingSinglecellAnalysis2020}{2}{]}: we normalize UMI counts by the total number of counts per cell, apply log-transformation and identify highly variable genes (HVGs). Then we project the \(H \times L\) gene expression matrix, where \(L\) is the number of cells and \(H\) is the number of HVGs, to the first \(d\) principal components (PCs). While downstream analysis is generally robust to the exact choice of the number of HVGs {[}\protect\hyperlink{ref-lueckenCurrentBestPractices2019}{1}{]}, an optimal value for \(d\) can be selected by detecting the ``elbow'' in the variance explained by PCs or using the ``jackstraw'' method {[}\protect\hyperlink{ref-chungStatisticalSignificanceVariables2015}{3}{]}.

\hypertarget{minimizing-batch-effects}{%
\subsubsection{Minimizing batch effects}\label{minimizing-batch-effects}}

Comparing biological conditions often requires acquiring single-cell data from multiple samples, that can be generated with different experimental conditions or protocols. This commonly introduces batch effects, which can have a substantial impact on the data composition and subsequently the topology of any KNN graph computed across the single-cell data. Consequently, this will have an impact on the ability of \emph{Milo} to resolve genuine differential abundance of cells between experimental conditions of interest. In addition, other biological nuisance covariates can impact DA analysis i.e.~biological factors that are not of interest for the analyst, such as donor of origin or sex of the donor. We recommend mitigating the impact of technical or other nuisance covariates \emph{before} building the KNN graph, by using one of the many \emph{in silico} integration tools designed for this task in single-cell datasets.
Defining the best tool for this task is beyond the scope of this work; we refer the reader to a large number of integration methods that have been reviewed and benchmarked in {[}\protect\hyperlink{ref-lueckenBenchmarkingAtlaslevelData2020}{4}--\protect\hyperlink{ref-tranBenchmarkBatcheffectCorrection2020}{6}{]}. However, users should consider the type of output produced by their integration method of choice, typically one of (A) a corrected feature space, (B) a joint embedding or (C) an integrated graph. The refined neighbourhood search procedure in \emph{Milo} relies on finding neighbors in reduced dimension space. Therefore using a batch-correction method that produces an integrated graph (e.g.~BBKNN {[}\protect\hyperlink{ref-polanskiBBKNNFastBatch}{7}{]}, Conos {[}\protect\hyperlink{ref-barkasJointAnalysisHeterogeneous2019}{8}{]}) may lead to sub-optimal results in DA testing with \emph{Milo}, as the refined neighbourhood search procedure would still be affected by the batch effect.

In addition, the effect of nuisance covariates should be modelled in the generalized linear model used for DA testing in \emph{Milo} to minimize the emergence of false positives in case of imperfect batch correction (see Section \ref{testDA}) (Fig 2D, Supp Fig 11).

We wish to emphasize that, in the presence of confounding factors, an appropriate experimental design is crucial to obtain reliable results from differential abundance analysis: if nuisance factors are 100\% confounded with the biological condition used for differential abundance (e.g.~if the samples from diseased and healthy donors are processed in separate sequencing batches), there is no way to disentangle the abundance differences that are truly driven by the biology of interest. In a similar case applying a batch integration strategy before graph construction could lead to a loss of biological signal.

\hypertarget{building-the-knn-graph}{%
\subsubsection{Building the KNN graph}\label{building-the-knn-graph}}

\emph{Milo} uses a KNN graph computed based on similarities in gene expression space as a representation of the phenotypic manifold in which cells lie. While \emph{Milo} can be used on graphs built with different similarity kernels, here we compute the graph as follows: given the reduced dimension matrix \(X_{PC}\) of dimensions \(L \times d\), for each cell \(c_j\), the Euclidean distances to its \(K\) nearest neighbors in \(X_{PC}\) are computed and stored in a \(L \times L\) adjacency matrix \(D\). Then, \(D\) is made symmetrical, such that cells \(c_i\) and \(c_j\) are nearest neighbors (i.e.~connected by an edge) if either \(c_i\) is a nearest neighbor of \(c_j\) or \(c_j\) is a nearest neighbor of \(c_i\). The KNN graph is encoded by the undirected symmetric version \(\tilde{D}\) of \(D\), where each cell has at least \(K\) nearest neighbors.

\hypertarget{nh}{%
\subsubsection{Definition of cell neighbourhoods and index sampling algorithm}\label{nh}}

Next, we identify a set of representative cell neighbourhoods on the KNN graph. We define the neighbourhood \(n_i\) of cell \(c_i\) as the group of cells that are connected to \(c_i\) by an edge in the graph. We refer to \(c_i\) with \(i=1,2,...,N\) as the index cell of the neighbourhood, so that \(N \leq L\).
Formally, a cell \(c_j\) belongs to neighbourhood \(n_i\) if \(\tilde{D}_{i,j} > 0\).

In order to define neighbourhoods that span the whole KNN graph, we sample index cells by using an algorithm previously adopted for waypoint sampling for trajectory inference {[}\protect\hyperlink{ref-gutTrajectoriesCellcycleProgression2015}{9},\protect\hyperlink{ref-settyWishboneIdentifiesBifurcating2016}{10}{]}.
Briefly, we start by randomly sampling \(p \cdot L\) cells from the dataset, where \(p \in [0,1]\) (we use \(p = 0.1\) by default).
Given the reduced dimension matrix used for graph construction \(X_{PC}\), for each sampled cell \(c_j\) we consider its \(K\) nearest neighbors with PC profiles \({x_1, x_2, ... , x_k}\) and compute the mean position of the neighbors in PC space \(\bar{x}\):
\[
\bar{x_j} = \frac{ \sum_k x_k }{K}
\]

Then, we search for the cell \(c_i\) such that the Euclidean distance between \(x_i\) and \(\bar{x_j}\) is minimized. Because the algorithm might converge to the same index cell from multiple initial samplings, this procedure yields a set of \(N \leq p \cdot L\) index cells that are used to define neighbourhoods.

Having defined a set of \(N\) neighbourhoods from the sampled index cells, we construct a count matrix of dimensions \(N \times S\) which reports, for each sample, the number of cells that are present in each neighbourhood.

\hypertarget{testDA}{%
\subsubsection{Testing for differential abundance in neighbourhoods}\label{testDA}}

To test for differential abundance between biological conditions, \emph{Milo} models the cell counts in neighbourhoods, estimating variability across biological replicates using a generalized linear model (GLM). We build upon the framework for differential abundance testing implemented by \emph{Cydar} {[}\protect\hyperlink{ref-lunTestingDifferentialAbundance2017}{11}{]}. In this section, we briefly describe the statistical model and adaptations to the KNN graph setting.

\hypertarget{quasi-likelihood-negative-binomial-generalized-linear-models}{%
\paragraph*{Quasi-likelihood negative binomial generalized linear models}\label{quasi-likelihood-negative-binomial-generalized-linear-models}}
\addcontentsline{toc}{paragraph}{Quasi-likelihood negative binomial generalized linear models}

We consider a neighbourhood \(n\) with cell counts \(y_{ns}\) for each experimental sample \(s\). The counts are modelled by the negative binomial (NB) distribution, as it is supported over all non-negative integers and can accurately model both small and large cell counts. For such non-Normally distributed data we use generalized-linear models (GLMs) as an extension of classic linear models that can accomodate complex experimental designs. We therefore assume that
\[
y_{ns} \sim NB(\mu_{ns},\phi_{n}),
\]
where \(\mu_{ns}\) is the mean number of cells from sample \(s\) in neighbourhood \(n\) and \(\phi_{n}\) is the NB dispersion parameter.

The expected count value \(\mu_{ns}\) is given by
\[
\mu_{ns} = \lambda_{ns}L_{s}
\]

where \(\lambda_{ns}\) is the proportion of cells belonging to experimental sample \(s\) in \(n\) and \(L_s\) is the sum of counts of cells of \(s\) over all the neighbourhoods. In practice, \(\lambda_{ns}\) represents the biological variability that can be affected by treatment condition, age or any biological covariate of interest.

We use a log-linear model to model the influence of a biological condition on the expected counts in the neighbourhood:

\begin{equation}
  \label{eq:1}
    log\ \mu_{ns} = \sum_{g=1}^{G}x_{sg}\beta_{ng} + log\ L_s
\end{equation}

Here, for each possible value \(g\) taken by the biological condition of interest, \(x_{sg}\) is the vector indicating the condition value applied to sample \(s\). \(\beta_{ng}\) is the regression coefficient by which the covariate effects are mediated for neighbourhood \(n\), that represents the log fold-change between number of cells in condition \(g\) and all other conditions. If the biological condition of interest is ordinal (such as age or disease-severity) \(\beta_{ng}\) is interpreted as the per-unit linear change in neighbourhood abundance.

Estimation of \(\beta_{ng}\) for each \(n\) and \(g\) is performed by fitting the GLM to the count data for each neighbourhood, i.e.~by estimating the dispersion \(\phi_{n}\) that models the variability of cell counts for replicate samples for each neighbourhood. Dispersion estimation is performed using the quasi-likelihood method in \texttt{edgeR}{[}\protect\hyperlink{ref-robinsonEdgeRBioconductorPackage2010a}{12}{]}, where the dispersion is modelled from the GLM deviance and thereby stabilized with empirical Bayes shrinkage, to stabilize the estimates in the presence of limited replication.

\hypertarget{count-model-normalisation-and-compositional-biases}{%
\paragraph*{Count model normalisation and compositional biases}\label{count-model-normalisation-and-compositional-biases}}
\addcontentsline{toc}{paragraph}{Count model normalisation and compositional biases}

In equation (\ref{eq:1}) above the \(\mbox{log}\ L_s\) term is provided as an offset to the NB GLM which effectively normalises the cell counts in
each neighbourhood by the total number of cells in each sample \(S\), thus accounting for variation in cell numbers across samples. If there is a
single strong region of differential abundance then the counts for these samples will increase, which can negatively bias the model log fold-change
estimates. This results in an underestimate of the true log fold-changes and the appearance of false discoveries in the opposite direction to the
true DA effect direction. To address this issue we
turn to the RNA-seq literature, specifically the trimmed mean of M-values (TMM) method for estimating normalisation factors that
are robust to such compositional differences across samples {[}\protect\hyperlink{ref-robinsonTMM2010}{13}{]}. Under the assumption that the majority of neighbourhoods are
not differentially abundant, the TMM approach first computes the per-neighbourhood log count ratios for a pair of samples \(s\) and \(s'\)
(M values):

\[ 
M_{n} = \mbox{log} \frac{y_{ns}/M_s}{y_{ns'}/M_{s'}}
\]

And the absolute neighbourhood abundance (A values):

\[
A_{n} = \frac{1}{2}\ \mbox{log}_{2} (y_{ns}/M_{s} \cdot y_{ns'}/M_{s'}),\ \mbox{for}\ y_{n} \neq 0
\]

Both the M and A distribution tails are trimmed (30\% for M, 5\% for A by default) before taking a weighted average over neighbourhoods using precision weights,
computed as the inverse variance of the neighbourhood counts, to account for the fact that more abundant neighbourhoods have a lower variance on a
log scale. Thus, the normalisation factors are computed, with respect to a reference sample, \(r\):

\[
\mbox{log}_2(TMM_{s}^{(r)})= \frac{\sum_{n \in N} w^r_{ns}M^r_{ns}}{\sum_{n \in N} w^r_{ns}}
\]

where, \(M^r_{ns}\) is computed as above for samples \(s\) and \(r\), and:

\[
w^r_{ns} = \frac{M_s-y_{ns}}{M_sy_{ns}} + \frac{M_r-y_{nr}}{M_ry_{nr}}
\]

In practice, \(M_r\) and \(y_{nr}\) are computed from the sample with the counts per million upper quartile that is closest to the mean upper quartile across samples.

\hypertarget{adaptation-of-spatial-fdr-to-neighbourhoods}{%
\paragraph*{Adaptation of Spatial FDR to neighbourhoods}\label{adaptation-of-spatial-fdr-to-neighbourhoods}}
\addcontentsline{toc}{paragraph}{Adaptation of Spatial FDR to neighbourhoods}

To control for multiple testing, we need to account for the overlap between neighbourhoods, that makes the differential abundance tests non-independent. We apply a weighted version of the Benjamini-Hochberg (BH) method, where p-values are weighted by the reciprocal of the neighbourhood connectivity, as an adaptation to graphs of the Spatial FDR method introduced by \(Cydar\) {[}\protect\hyperlink{ref-lunTestingDifferentialAbundance2017}{11}{]}.
Formally, to control for FDR at a selected threshold \(\alpha\) we reject null hypothesis \(i\) where the associated p-value is less than the threshold:

\[
\max_i{p_{(i)}: p_{(i)}\le \alpha\frac{\sum_{l=1}^{i}w_{(l)}}{\sum_{l=1}^{n}w_{(l)}}}
\]
Where the weight \(w_{(i)}\) is the reciprocal of the neighbourhood connectivity \(c_i\). As a measure of neighbourhood connectivity, we use the Euclidean distance between the neighbourhood index cell \(c_i\) and its kth nearest neighbour in PC space.

\hypertarget{guidelines-on-parameter-choice}{%
\subsection{Guidelines on parameter choice}\label{guidelines-on-parameter-choice}}

In this section we provide practical guidelines to select default parameters for KNN graph and neighbourhood construction for DA analysis with Milo. We recognize that DA analysis will also be impacted by choices made during feature selection and dimensionality reduction. However these depend strongly on the nature of the single-cell dataset used as input. For example feature selection strategies suitable for UMI-based scRNA-seq data might be suboptimal for data generated with non-UMI protocols, or dimensionality reduction methods alternative to PCA might be used for single-cell epigenomics data. We point the reader to existing resources and heuristics for the application to scRNA-seq in section \ref{pp}.

\hypertarget{selecting-the-number-of-nearest-neighbors-k}{%
\paragraph*{\texorpdfstring{Selecting the number of nearest neighbors \(K\)}{Selecting the number of nearest neighbors K}}\label{selecting-the-number-of-nearest-neighbors-k}}
\addcontentsline{toc}{paragraph}{Selecting the number of nearest neighbors \(K\)}

For construction of the KNN graph and neighbourhoods, the user has to select the number of nearest neighbors \(K\) to use for graph construction.
The choice of \(K\) influences the distribution of cell counts within neighbourhoods, as \(K\) represents the lower limit in the number of cells in each neighbourhood (\(\sum(y_{n,s})\)). Hence, if \(K\) is too small the neighbourhoods might not contain enough cells to detect differential abundance. As we illustrate by testing for DA with increasing values for \(K\) in the mouse gastrulation dataset with synthetic condition labels (Supp Note Fig 1A-B) increasing \(K\) increases power, but can come at the cost of FDR control.
In order to perform DA testing with sufficient statistical power, the analyst should consider the number of experimental samples \(S\) (that will correspond to the columns in the count matrix for DA testing) and the desired minimum number of cells per neighbourhood and experimental sample.
The median number of cells per sample in each neighbourhood \(\hat{y}_{ns}\) increases with the total neighbourhood size (Supp Note Fig 1C), with:

\[
\hat{y}_{ns} \sim \frac{\sum_s y_{ns}}{S}
\]
Therefore a conservative approach to minimize false positives is to select \(K \geq S \times\) 3-5.

We recommend users to inspect the histogram of neighbourhood sizes after sampling of neighbourhoods (Supp Note Fig 1D) and to consider the number of cells that would be considered a ``neighbourhood'' in the dataset at hand. As a heuristic for selecting a lower bound on \(K\) to increase the resolution of neighbourhoods for capturing rare sub-populations or states, the user can select \(K\) such that the mean neighbourhood size is no more than 10\% of the expected size of the rare population. We provide the utility function
\texttt{plotNhoodSizeHist} to visualize the neighbourhood size distribution as part of our R package.

To verify how robust the findings from Milo are to the choice of K, we repeated DA analysis on both the mouse thymus and human liver data sets presented in the Results across a range of values of K, from very small (K=2) to very large (K=100). We found that for these 2 data sets the DA regions correspond to those identified in Fig 4-5 across a range of values of K (Supp Note Fig 2). We computed the log fold change for the DA neighbourhoods at each value of K and found that the differential abundance results are robust, with loss of power seen only at very small values (K\textless10). As K becomes very large (K\textgreater50), neighborhoods contain more heterogeneous mixtures of cells, leading to an ``over-smoothing'' and a loss of resolution for rarer cell states in the thymus dataset (for example in the sTEC cluster).

\hypertarget{selecting-the-proportions-of-cells-sampled-as-neighbourhood-indices-p}{%
\paragraph*{\texorpdfstring{Selecting the proportions of cells sampled as neighbourhood indices \(p\)}{Selecting the proportions of cells sampled as neighbourhood indices p}}\label{selecting-the-proportions-of-cells-sampled-as-neighbourhood-indices-p}}
\addcontentsline{toc}{paragraph}{Selecting the proportions of cells sampled as neighbourhood indices \(p\)}

The proportion of cells sampled for search of neighbourhood indices can affect the total number of neighbourhoods used for analysis, but this number will converge for high proportions thanks to the sampling refinement step described in section \ref{nh} (Supp Fig 1A). In practice, we recommend initiating neighbourhood search with \(p=0.05\) for datasets with more than 100k cells and \(p=0.1\) otherwise, which we have found to give appropriate coverage across the KNN graph while reducing the computational and multiple-testing burden. We recommend selecting \(p > 0.1\) only if the dataset appears to contain rare disconnected subpopulations.

\hypertarget{notes-on-experimental-design}{%
\subsection{Notes on experimental design}\label{notes-on-experimental-design}}

Of key consideration when designing any single-cell experiment is how the sample collection relates to the biological variables of interest, and how these samples
are processed and experiments are performed. Moreover, the experimenter (and analyst together), should design their experiment to minimise the impact of confounding
effects on differential abundance testing, and incorporate appropriate replication to acheive enough power to detect the expected effect size for their
experiment.

\hypertarget{statistical-power-considerations}{%
\paragraph*{Statistical power considerations}\label{statistical-power-considerations}}
\addcontentsline{toc}{paragraph}{Statistical power considerations}

Increases in statistical power can be achieved by several means: (1) Increased cell numbers in neighbourhoods and (2) higher signal-to-noise ratio. The first can
be achieved by collecting more cells for each sample, increasing \(K\) during graph building such that neighbourhoods are on average larger, and by increasing the
number of replicate samples. Collecting more cells gives a greater coverage of the cell-to-cell heterogeneity and different cell states/types, including increased
detection for rarer sub-populations. Increasing \(K\) increases power by constructing larger neighbourhoods, however, this increase in power comes at a cost of
reduced sensitivity for rarer sub-populations and an increased false discovery rate (Supp Note Fig 1A-B). Designing an experiment with more replicate samples has
multiple benefits in terms of increasing statistical testing power, increasing the signal-to-noise ratio, and increasing the accuracy of effect size estimates
(Supp Fig 8B). Therefore, in order of their impact on power and differential abundance testing, we would recommend: (1) collecting more replicate samples, with
a minimum of n=3, (2) collecting more cells per sample, (3) increasing \(K\) to generate larger neighbourhoods.

\hypertarget{batch-effects-and-experimental-design}{%
\paragraph*{Batch effects and experimental design}\label{batch-effects-and-experimental-design}}
\addcontentsline{toc}{paragraph}{Batch effects and experimental design}

Proper experimental design is crucial for answering scientific questions, particularly in the presence of confounding effects. In single-cell experiments these
can range from batch effects introduced between samples processed on different days, owing to logistical constraints or sample availability, to biological sample
collections from a heterogenous population; the latter being particularly apparent for genetically diverse non-model organisms.

In the context of differential abundance testing with Milo, we recommend designing experimental procedures and sample processing such that samples from different
conditions are randomised across batches. One example is to pair samples between conditions, such that during batch effect removal the variability between these
pairs of samples is minimally removed. This will help to facilitate removal of technical batch effects, whilst retaining the relevant biological variability.\\
As described above, the exact choice of batch integration method should be carefully considered before applying Milo, with a preference for methods that generate
a batch-integrated space (either reduced dimensions or gene expression). The key point is that sample processing and experimental batches are not perfectly
confounded with the biological variable of interest. We expect \emph{some} technical variability to remain (no batch integration is perfect), which can be handled in
Milo's GLM framework by including the batch identity as a blocking factor in the design model. Examples of this correction are shown in the benchmarking in
Fig 2E and supp Fig 11.

\hypertarget{supplementary-note-figures}{%
\section{Supplementary Note Figures}\label{supplementary-note-figures}}

\renewcommand{\figurename}{Supplementary Note Figure}
\pagenumbering{gobble}

\begin{figure}
\centering
\includegraphics{suppl_figs/suppl_fig_Kselection.pdf}
\caption{\label{fig:sup-fig-Kselection}\textbf{Selection of \(K\) parameter}
(A-B) Example trends for TPR and FDR for increasing values of K used for KNN graph building on simulated DA on 8 regions (\(P(C1) = 0.8\)). Dotted lines highlight TPR=0.8 and FDR=0.1 thresholds. (C) The median number of cells per experimental sample is a function of the neighbourhood size \(\sum_s{y_{n,s}}\) divided by the total number of samples \(S\). (D-F) Histogram of neighbourhood sizes for different choices of K. The red dotted line denotes the minimum neighbourhood size to obtain 5 cells per sample on average.}
\end{figure}




\begin{figure}
\centering
\includegraphics{suppl_figs/suppl_fig_krobustness.pdf}
\caption{\label{fig:sup-fig-robustness}\textbf{Robustness of Milo DA testing to varying K}. Distributions of DA neighbourhoods across values of K for the mouse ageing thymus (A) and human cirrhotic liver (B) data sets. Shown are the distributions of log fold-changes (y-axis) for DA (FDR 10\%) neighbourhoods using different values of K (x-axis) from 5-100, illustrating that DA testing is robust across a broad range of values of K.}
\end{figure}



\newpage

\hypertarget{references}{%
\section*{References}\label{references}}
\addcontentsline{toc}{section}{References}

\hypertarget{refs}{}
\begin{CSLReferences}{0}{0}
\leavevmode\hypertarget{ref-lueckenCurrentBestPractices2019}{}%
\CSLLeftMargin{1. }
\CSLRightInline{Luecken, M.D., and Theis, F.J. (2019). Current best practices in single-cell {RNA}-seq analysis: A tutorial. Molecular Systems Biology \emph{15}, e8746.}

\leavevmode\hypertarget{ref-amezquitaOrchestratingSinglecellAnalysis2020}{}%
\CSLLeftMargin{2. }
\CSLRightInline{Amezquita, R.A., Lun, A.T.L., Becht, E., Carey, V.J., Carpp, L.N., Geistlinger, L., Marini, F., Rue-Albrecht, K., Risso, D., Soneson, C., \emph{et al.} (2020). Orchestrating single-cell analysis with {Bioconductor}. Nature Methods \emph{17}, 137--145.}

\leavevmode\hypertarget{ref-chungStatisticalSignificanceVariables2015}{}%
\CSLLeftMargin{3. }
\CSLRightInline{Chung, N.C., and Storey, J.D. (2015). Statistical significance of variables driving systematic variation in high-dimensional data. Bioinformatics \emph{31}, 545--554.}

\leavevmode\hypertarget{ref-lueckenBenchmarkingAtlaslevelData2020}{}%
\CSLLeftMargin{4. }
\CSLRightInline{Luecken, M.D., Büttner, M., Chaichoompu, K., Danese, A., Interlandi, M., Mueller, M.F., Strobl, D.C., Zappia, L., Dugas, M., Colomé-Tatché, M., \emph{et al.} (2020). Benchmarking atlas-level data integration in single-cell genomics. bioRxiv, 2020.05.22.111161.}

\leavevmode\hypertarget{ref-chazarra-gilFlexibleComparisonBatch2020}{}%
\CSLLeftMargin{5. }
\CSLRightInline{Chazarra-Gil, R., Dongen, S. van, Kiselev, V.Y., and Hemberg, M. (2020). Flexible comparison of batch correction methods for single-cell {RNA}-seq using {BatchBench}. bioRxiv, 2020.05.22.111211.}

\leavevmode\hypertarget{ref-tranBenchmarkBatcheffectCorrection2020}{}%
\CSLLeftMargin{6. }
\CSLRightInline{Tran, H.T.N., Ang, K.S., Chevrier, M., Zhang, X., Lee, N.Y.S., Goh, M., and Chen, J. (2020). A benchmark of batch-effect correction methods for single-cell {RNA} sequencing data. Genome Biology \emph{21}, 12.}

\leavevmode\hypertarget{ref-polanskiBBKNNFastBatch}{}%
\CSLLeftMargin{7. }
\CSLRightInline{Polański, K., Young, M.D., Miao, Z., Meyer, K.B., Teichmann, S.A., and Park, J.-E. {BBKNN}: Fast batch alignment of single cell transcriptomes. Bioinformatics.}

\leavevmode\hypertarget{ref-barkasJointAnalysisHeterogeneous2019}{}%
\CSLLeftMargin{8. }
\CSLRightInline{Barkas, N., Petukhov, V., Nikolaeva, D., Lozinsky, Y., Demharter, S., Khodosevich, K., and Kharchenko, P.V. (2019). Joint analysis of heterogeneous single-cell {RNA}-seq dataset collections. Nat Methods \emph{16}, 695--698.}

\leavevmode\hypertarget{ref-gutTrajectoriesCellcycleProgression2015}{}%
\CSLLeftMargin{9. }
\CSLRightInline{Gut, G., Tadmor, M.D., Pe'er, D., Pelkmans, L., and Liberali, P. (2015). Trajectories of cell-cycle progression from fixed cell populations. Nature Methods \emph{12}, 951--954.}

\leavevmode\hypertarget{ref-settyWishboneIdentifiesBifurcating2016}{}%
\CSLLeftMargin{10. }
\CSLRightInline{Setty, M., Tadmor, M.D., Reich-Zeliger, S., Angel, O., Salame, T.M., Kathail, P., Choi, K., Bendall, S., Friedman, N., and Pe'er, D. (2016). Wishbone identifies bifurcating developmental trajectories from single-cell data. Nature Biotechnology \emph{34}, 637--645.}

\leavevmode\hypertarget{ref-lunTestingDifferentialAbundance2017}{}%
\CSLLeftMargin{11. }
\CSLRightInline{Lun, A.T.L., Richard, A.C., and Marioni, J.C. (2017). Testing for differential abundance in mass cytometry data. Nature Methods \emph{14}, 707--709.}

\leavevmode\hypertarget{ref-robinsonEdgeRBioconductorPackage2010a}{}%
\CSLLeftMargin{12. }
\CSLRightInline{Robinson, M.D., McCarthy, D.J., and Smyth, G.K. (2010). {edgeR}: A {Bioconductor} package for differential expression analysis of digital gene expression data. Bioinformatics \emph{26}, 139--140.}

\leavevmode\hypertarget{ref-robinsonTMM2010}{}%
\CSLLeftMargin{13. }
\CSLRightInline{Robinson, M.D., and Oshlack, A. (2010). A scaling normalization method for differential expression analysis of {RNA}-seq data. Genome Biology \emph{11}, R25. Available at: \url{http://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25} {[}Accessed March 18, 2021{]}.}

\end{CSLReferences}

\end{document}
