% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\hypersetup{
  pdftitle={Supplementary information for Differential cell-state abundance testing using k-NN graphs with Milo},
  pdfauthor={Emma Dann,; Neil C. Henderson,; Sarah A. Teichmann,; Michael D. Morgan,; John C. Marioni},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}
\urlstyle{same} % disable monospaced font for URLs
\usepackage[margin=1in]{geometry}
\usepackage{longtable,booktabs}
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
\usepackage{lineno}
\linenumbers
\renewcommand{\baselinestretch}{1.8}
\ifluatex
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\newlength{\cslhangindent}
\setlength{\cslhangindent}{1.5em}
\newenvironment{cslreferences}%
  {}%
  {\par}

\title{Supplementary information for
\textbf{Differential cell-state abundance testing using k-NN graphs with \emph{Milo}}}
\author{Emma Dann, \and Neil C. Henderson, \and Sarah A. Teichmann, \and Michael D. Morgan, \and John C. Marioni}
\date{08 March, 2021}

\begin{document}
\maketitle

{
\setcounter{tocdepth}{3}
\tableofcontents
}
\renewcommand{\figurename}{Supplementary Figure}

\newpage

\hypertarget{supplementary-notes}{%
\section{Supplementary notes}\label{supplementary-notes}}

\hypertarget{description-of-workflow-for-milo-analysis}{%
\subsection{\texorpdfstring{Description of workflow for \emph{Milo} analysis}{Description of workflow for Milo analysis}}\label{description-of-workflow-for-milo-analysis}}

Given a single-cell dataset of gene expression profiles of \(M\) cells collected from \(S\) experimental samples, \emph{Milo} aims to quantify systematic changes in abundance of cells between biological conditions, as compared to within-condition variability. Here we provide a step-by-step description of the workflow for differential abundance analysis. Of note, here we focus on application to single-cell gene expression profiles, and we provide guidelines for preprocessing on this type of data. However, the core of the \emph{Milo} framework, from k-NN graph construction to differential abundance testing, is applicable to any kind of single-cell dataset that can be embedded in a low-dimensional space.

\hypertarget{pp}{%
\subsubsection{Preprocessing and dimensionality reduction}\label{pp}}

For preprocessing of scRNA-seq profiles we recommend following standard practices in single-cell analysis {[}\protect\hyperlink{ref-lueckenCurrentBestPractices2019}{1},\protect\hyperlink{ref-amezquitaOrchestratingSinglecellAnalysis2020}{2}{]}: we normalize UMI counts by the total number of counts per cell, apply log-transformation and identify highly variable genes (HVGs). Then we project the \(H \times M\) gene expression matrix, where \(M\) is the number of cells and \(H\) is the number of HVGs, to the first \(d\) principal components (PCs). While downstream analysis is generally robust to the exact choice of the number of HVGs {[}\protect\hyperlink{ref-lueckenCurrentBestPractices2019}{1}{]}, an optimal value for \(d\) can be selected by detecting the ``elbow'' in the variance explained by PCs or using the ``jackstraw'' method {[}\protect\hyperlink{ref-chungStatisticalSignificanceVariables2015}{3}{]}.

\hypertarget{minimizing-batch-effects}{%
\subsubsection{Minimizing batch effects}\label{minimizing-batch-effects}}

Comparing biological conditions often requires acquiring single-cell data from multiple samples, that can be generated with different experimental conditions or protocols. This commonly introduces batch effects, which can have a substantial impact on the data composition and subsequently the topology of any k-NN graph computed across the single-cell data. Consequently, this will have an impact on the ability of \emph{Milo} to resolve genuine differential abundance of cells between experimental conditions of interest. In addition, other biological nuisance covariates could impact DA analysis i.e.~biological factors that are not of interest for the analyst, such as donor of origin or sex of the donor. We recommend to mitigate the impact of technical or other nuisance covariates \emph{before} building the k-NN graph, by using one of the many \emph{in silico} integration tools designed for this task in single-cell datasets.
Defining the best tool for this task is beyond the scope of this work, a large number of integration methods have been reviewed and benchmarked in {[}\protect\hyperlink{ref-lueckenBenchmarkingAtlaslevelData2020}{4}--\protect\hyperlink{ref-tranBenchmarkBatcheffectCorrection2020}{6}{]}. However, users should consider the type of output produced by their integration method of choice, typically one of (A) a corrected feature space, (B) a joint embedding or (C) an integrated graph. The refined neighbourhood search procedure in \emph{Milo} relies on finding neighbors in reduced dimension space. Therefore using a methods that produces an integrated graph (e.g.~BBKNN {[}\protect\hyperlink{ref-polanskiBBKNNFastBatch}{7}{]}, Conos {[}\protect\hyperlink{ref-barkasJointAnalysisHeterogeneous2019}{8}{]}) could lead to suboptimal results in DA testing with \emph{Milo}, because the refined neighbourhood search procedure would still be affected by the batch effect.

In addition, the effect of nuisance covariates should be modelled in the generalized linear model used for DA testing in \emph{Milo} to minimize the emergence of false positives in case of imperfect batch correction (see Section \ref{testDA}) (Fig.2D).

We wish to emphasize that, when confounders are present, an appropriate experimental design is crucial to obtain reliable results from differential abundance analysis: if nuisance factors are 100\% confounded with the biological condition used for differential abundance (e.g.~if the samples from diseased and healthy donors are processed in separate sequencing batches), there is no way to disentangle the abundance differences that are truly driven by the biology of interest. In a similar case applying a batch integration strategy before graph construction could lead to a loss of biological signal.

\hypertarget{building-the-k-nn-graph}{%
\subsubsection{Building the k-NN graph}\label{building-the-k-nn-graph}}

\emph{Milo} uses a k-NN graph computed based on similarities in gene expression space as a representation of the phenotypic manifold in which cells lie. While \emph{Milo} can be used on graphs built with different similarity kernels, here we compute the graph as follows: given the reduced dimension matrix \(X_{PC}\) of dimensions \(M \times d\), for each cell \(j\), the Euclidean distances to its \(K\) nearest neighbors in \(X_{PC}\) are computed and stored in a \(M \times M\) adjacency matrix \(D\). Then, \(D\) is made symmetrical, such that cells \(c_i\) and \(c_j\) are nearest neighbors (i.e.~connected by an edge) if either \(c_i\) is a nearest neighbor of \(c_j\) or \(c_j\) is a nearest neighbor of \(c_i\). The k-NN graph is encoded by the undirected symmetric version of \(\tilde{D}\) of \(D\), where each cell has at least \(K\) nearest neighbors.

\hypertarget{nh}{%
\subsubsection{Definition of cell neighbourhoods and index sampling algorithm}\label{nh}}

Next, we identify a set of representative cell neighbourhoods on the k-NN graph. We define the neighbourhood \(n_i\) of cell \(c_i\) as the group of cells that are connected to \(c_i\) by an edge in the graph. We refer to \(c_i\) with \(i=1,2,...,N\) as the index cell of the neighbourhood, so that \(N \leq M\).
Formally, a cell \(c_j\) belongs to neighbourhood \(n_i\) if \(\tilde{D}_{i,j} > 0\).

In order to define neighbourhoods that span the whole k-NN graph, we sample index cells by using an algorithm previously adopted for waypoint sampling for trajectory inference {[}\protect\hyperlink{ref-gutTrajectoriesCellcycleProgression2015}{9},\protect\hyperlink{ref-settyWishboneIdentifiesBifurcating2016}{10}{]}.
Briefly, we start by randomly sampling \(p \cdot M\) cells from the dataset, where \(p \in [0,1]\) (we use \(p = 0.1\) by default).
Given the reduced dimension matrix used for graph construction \(X_{PC}\), for each sampled cell \(c_j\) we consider its \(K\) nearest neighbors with PC profiles \({x_1, x_2, ... , x_k}\) and compute the mean position of the neighbors in PC space \(\bar{x}\):
\[
\bar{x_j} = \frac{ \sum_k x_k }{K}
\]

Then, we search for the cell \(c_i\) such that the Euclidean distance between \(x_i\) and \(\bar{x}\) is minimized. Because the algorithm might converge to the same index cell from multiple initial samplings, this procedure yields a set of \(N \leq p \cdot M\) index cells that are used to define neighbourhoods.

Having defined a set of \(N\) neighbourhoods from the sampled index cells, we construct a count matrix of dimensions \(N \times S\) which reports, for each sample, the number of cells that are present in each neighbourhood.

\hypertarget{testDA}{%
\subsubsection{Testing for differential abundance in neighbourhoods}\label{testDA}}

To test for differential abundance between biological conditions, \emph{Milo} models the cell counts in neighbourhoods, estimating variability across biological replicates using a generalized linear model (GLM). We build upon the framework for differential abundance testing implemented by \emph{Cydar} {[}\protect\hyperlink{ref-lunTestingDifferentialAbundance2017}{11}{]}. In this section, we briefly describe the statistical model and adaptations to the k-NN graph setting.

\hypertarget{quasi-likelihood-negative-binomial-generalized-linear-models}{%
\paragraph*{Quasi-likelihood negative binomial generalized linear models}\label{quasi-likelihood-negative-binomial-generalized-linear-models}}
\addcontentsline{toc}{paragraph}{Quasi-likelihood negative binomial generalized linear models}

We consider a neighbourhood \(n\) with cell counts \(y_{ns}\) for each experimental sample \(s\). The counts are modelled by the negative binomial (NB) distribution, as it is supported over all non-negative integers and can accurately model both small and large cell counts. For such non-Normally distributed data we use generalized-linear models (GLMs) as an extension of classic linear models that can accomodate complex experimental designs. We therefore assume that
\[
y_{ns} \sim NB(\mu_{ns},\phi_{n}),
\]
where \(\mu_{ns}\) is the mean number of cells from sample \(s\) in neighbourhood \(n\) and \(\phi_{n}\) is the NB dispersion parameter.

The expected count value \(\mu_{ns}\) is given by
\[
\mu_{ns} = \lambda_{ns}M_{ns}
\]

where \(\lambda_{ns}\) is the proportion of cells belonging to experimental sample \(s\) in \(n\) and \(M_s\) is the total number of cells of \(s\). In practice, \(\lambda_{ns}\) represents the biological variability that can be affected by treatment condition, age or any biological covariate of interest.

We use a log-linear model to model the influence of a biological condition on the expected counts in the neighbourhood:
\[
log\ \mu_{ns} = \sum_{g=1}^{G}x_{sg}\beta_{ng} + log\ M_s
\]
Here, for each possible value \(g\) taken by the biological condition of interest, \(x_{sg}\) is the binary vector indicating the condition value applied to sample \(s\). \(\beta_{ng}\) is the regression coefficient by which the covariate effects are mediated for neighbourhood \(n\), that represents the log fold-change between number of cells in condition \(g\) and all other conditions. If the biological condition of interest is ordinal (such as age or disease-severity) \(\beta_{ng}\) is interpreted as the per-unit linear change in neighbourhood abundance.

Estimation of \(\beta_{ng}\) for each \(n\) and \(g\) is performed by fitting the GLM to the count data for each neighbourhood, i.e.~by estimating the dispersion \(\phi_{n}\) that models the variability of cell counts for replicate samples for each neighbourhood. Dispersion estimation is performed using the quasi-likelihood method in \texttt{edgeR}{[}\protect\hyperlink{ref-robinsonEdgeRBioconductorPackage2010a}{12}{]}, where the dispersion is modelled from the GLM deviance and thereby stabilized with empirical Bayes shrinkage, to stabilize the estimates in the presence of limited replication.

\hypertarget{adaptation-of-spatial-fdr-to-neighbourhoods}{%
\paragraph*{Adaptation of Spatial FDR to neighbourhoods}\label{adaptation-of-spatial-fdr-to-neighbourhoods}}
\addcontentsline{toc}{paragraph}{Adaptation of Spatial FDR to neighbourhoods}

To control for multiple testing, we need to account for the overlap between neighbourhoods, that makes the differential abundance tests non-independent. We apply a weighted version of the Benjamini-Hochberg (BH) method, where p-values are weighted by the reciprocal of the neighbourhood connectivity, as an adaptation to graphs of the Spatial FDR method introduced by \(Cydar\) {[}\protect\hyperlink{ref-lunTestingDifferentialAbundance2017}{11}{]}.
Formally, to control for FDR at a selected threshold \(\alpha\) we reject null hypothesis \(i\) where the associated p-value is less than the threshold:

\[
\max_i{p_{(i)}: p_{(i)}\le \alpha\frac{\sum_{l=1}^{i}w_{(l)}}{\sum_{l=1}^{n}w_{(l)}}}
\]
Where the weight \(w_{(i)}\) is the reciprocal of the neighbourhood connectivity \(c_i\). As a measure of neighbourhood connectivity, we use the Euclidean distance between the neighbourhood index cell \(c_i\) and its kth nearest neighbour in PC space.

\hypertarget{guidelines-on-parameter-choice}{%
\subsection{Guidelines on parameter choice}\label{guidelines-on-parameter-choice}}

In this section we provide practical guidelines to select default parameters for k-NN graph and neighbourhood construction for DA analysis with Milo. We recognize that DA analysis will also be impacted by choices made during feature selection and dimensionality reduction. However these depend strongly on the nature of the single-cell dataset used as input. For example feature selection strategies suitable for UMI-based scRNA-seq data might be suboptimal for data generated with non-UMI protocols, or dimensionality reduction methods alternative to PCA might be used for single-cell epigenomics data. We point the reader to existing resources and heuristics for the application to scRNA-seq in section \ref{pp}.

\hypertarget{selecting-the-number-of-nearest-neighbors-k}{%
\paragraph*{\texorpdfstring{Selecting the number of nearest neighbors \(K\)}{Selecting the number of nearest neighbors K}}\label{selecting-the-number-of-nearest-neighbors-k}}
\addcontentsline{toc}{paragraph}{Selecting the number of nearest neighbors \(K\)}

For construction of the k-NN graph and neighbourhoods, the user has to select the number of nearest neighbors \(K\) to use for graph construction.
The choice of \(K\) influences the distribution of cell counts within neighbourhoods, as \(K\) represents the lower limit in the number of cells in each neighbourhood (\(\sum(y_{n,s})\)). Hence, if \(K\) is too small the neighbourhoods might not contain enough cells to detect differential abundance. In order to perform DA testing with sufficient statistical power, the analyst should consider the number of experimental samples \(S\) (that will correspond to the columns in the count matrix for DA testing) and the desired minimum number of cells per neighbourhood and experimental sample. For example, having on average 5 cells per sample in each neighbourhood allows to detect n fold changes between 2 experimental conditions.
The median number of cells per sample in each neighbourhood \(\hat{y}_{ns}\) increases with the total neighbourhood size (Suppl.Fig\ldots.A), with:

\[
\hat{y}_{ns} \sim \frac{\sum_s y_{ns}}{S}
\]
Therefore a conservative approach to minimize false positives is to select \(K \geq S \times\) 3-5.

On the other hand increasing \(K\) increases power, but can come at the cost of FDR control, as we illustrate by testing for DA with increasing values for \(K\) in the mouse gastrulation dataset with synthetic condition labels on 4 different populations (Suppl.Fig\ldots.A). We recommend to inspect the histogram of neighbourhood sizes after sampling of neighbourhoods (Suppl.Fig\ldots.B) and to consider what is the number of cells that would be considered a ``neighbourhood'' in the dataset at hand. Depending on the expected lower-limit of the size of rare sub-populations, the mean neighbourhood size should be no more than 10\% of this value. We provide the utility function \texttt{plotNhoodSizeHist} to visualize the neighbourhood size distribution as part of our R package.

\hypertarget{selecting-the-proportions-of-cells-sampled-as-neighbourhood-indices-p}{%
\paragraph*{\texorpdfstring{Selecting the proportions of cells sampled as neighbourhood indices \(p\)}{Selecting the proportions of cells sampled as neighbourhood indices p}}\label{selecting-the-proportions-of-cells-sampled-as-neighbourhood-indices-p}}
\addcontentsline{toc}{paragraph}{Selecting the proportions of cells sampled as neighbourhood indices \(p\)}

The proportion of cells sampled for search of neighbourhood indices can affect the total number of neighbourhoods used for analysis, but this number will converge for high proportions thanks to the sampling refinement step described in section \ref{nh} (Suppl.Fig. 1A). In practice, we recommend initiating neighbourhood search with \(p=0.05\) for datasets with more than 100k cells and \(p=0.1\) otherwise, which we have found to give appropriate coverage across the k-NN graph while reducing the computational and multiple-testing burden. We recommend increasing \(p > 0.1\) only if the dataset appears to contain rare disconnected subpopulations.

\hypertarget{references}{%
\section*{References}\label{references}}
\addcontentsline{toc}{section}{References}

\hypertarget{refs}{}
\begin{cslreferences}
\leavevmode\hypertarget{ref-lueckenCurrentBestPractices2019}{}%
1. Luecken, M.D., and Theis, F.J. (2019). Current best practices in single-cell RNA-seq analysis: A tutorial. Molecular Systems Biology \emph{15}, e8746.

\leavevmode\hypertarget{ref-amezquitaOrchestratingSinglecellAnalysis2020}{}%
2. Amezquita, R.A., Lun, A.T.L., Becht, E., Carey, V.J., Carpp, L.N., Geistlinger, L., Marini, F., Rue-Albrecht, K., Risso, D., and Soneson, C. \emph{et al.} (2020). Orchestrating single-cell analysis with Bioconductor. Nature Methods \emph{17}, 137--145.

\leavevmode\hypertarget{ref-chungStatisticalSignificanceVariables2015}{}%
3. Chung, N.C., and Storey, J.D. (2015). Statistical significance of variables driving systematic variation in high-dimensional data. Bioinformatics \emph{31}, 545--554.

\leavevmode\hypertarget{ref-lueckenBenchmarkingAtlaslevelData2020}{}%
4. Luecken, M.D., Büttner, M., Chaichoompu, K., Danese, A., Interlandi, M., Mueller, M.F., Strobl, D.C., Zappia, L., Dugas, M., and Colomé-Tatché, M. \emph{et al.} (2020). Benchmarking atlas-level data integration in single-cell genomics. bioRxiv, 2020.05.22.111161.

\leavevmode\hypertarget{ref-chazarra-gilFlexibleComparisonBatch2020}{}%
5. Chazarra-Gil, R., Dongen, S. van, Kiselev, V.Y., and Hemberg, M. (2020). Flexible comparison of batch correction methods for single-cell RNA-seq using BatchBench. bioRxiv, 2020.05.22.111211.

\leavevmode\hypertarget{ref-tranBenchmarkBatcheffectCorrection2020}{}%
6. Tran, H.T.N., Ang, K.S., Chevrier, M., Zhang, X., Lee, N.Y.S., Goh, M., and Chen, J. (2020). A benchmark of batch-effect correction methods for single-cell RNA sequencing data. Genome Biology \emph{21}, 12.

\leavevmode\hypertarget{ref-polanskiBBKNNFastBatch}{}%
7. Polański, K., Young, M.D., Miao, Z., Meyer, K.B., Teichmann, S.A., and Park, J.-E. BBKNN: Fast batch alignment of single cell transcriptomes. Bioinformatics.

\leavevmode\hypertarget{ref-barkasJointAnalysisHeterogeneous2019}{}%
8. Barkas, N., Petukhov, V., Nikolaeva, D., Lozinsky, Y., Demharter, S., Khodosevich, K., and Kharchenko, P.V. (2019). Joint analysis of heterogeneous single-cell RNA-seq dataset collections. Nat Methods \emph{16}, 695--698.

\leavevmode\hypertarget{ref-gutTrajectoriesCellcycleProgression2015}{}%
9. Gut, G., Tadmor, M.D., Pe'er, D., Pelkmans, L., and Liberali, P. (2015). Trajectories of cell-cycle progression from fixed cell populations. Nature Methods \emph{12}, 951--954.

\leavevmode\hypertarget{ref-settyWishboneIdentifiesBifurcating2016}{}%
10. Setty, M., Tadmor, M.D., Reich-Zeliger, S., Angel, O., Salame, T.M., Kathail, P., Choi, K., Bendall, S., Friedman, N., and Pe'er, D. (2016). Wishbone identifies bifurcating developmental trajectories from single-cell data. Nature Biotechnology \emph{34}, 637--645.

\leavevmode\hypertarget{ref-lunTestingDifferentialAbundance2017}{}%
11. Lun, A.T.L., Richard, A.C., and Marioni, J.C. (2017). Testing for differential abundance in mass cytometry data. Nature Methods \emph{14}, 707--709.

\leavevmode\hypertarget{ref-robinsonEdgeRBioconductorPackage2010a}{}%
12. Robinson, M.D., McCarthy, D.J., and Smyth, G.K. (2010). edgeR: A Bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics \emph{26}, 139--140.
\end{cslreferences}

\end{document}
