---
output:
  bookdown::pdf_document2:
    keep_tex: yes
    toc: false
bibliography: milo_supplement.bib
csl: cell-numeric.csl
link-citations: true
geometry: margin=1cm
---

\renewcommand{\figurename}{Supplementary Figure}
\pagenumbering{gobble}

\newpage 

```{r sup-fig-refined, fig.pos="ht", fig.cap="(ref:sup-fig-refined)", echo=FALSE, out.width = "500px"}
knitr::include_graphics("suppl_figs/suppl_fig_refined.pdf", auto_pdf = TRUE)
```

(ref:sup-fig-refined) **Random sampling of k-NN graph vertices is suboptimal compared to sampling with refinement.**
(A) Sampling with refinement leads to selection of fewer neighbourhoods 
(B) Sampling with refinement leads to selection of bigger neighbourhoods for DA testing, independently of the initial proportion of cells sampled
(C) Sampling with refinement generates robust neighbourhoods across initializations: for each index cell we calculate the distance from the closest index in a sampling with different initialization. The cumulative distribution of distances to the closest index is shown. The black dotted line denotes the distribution of distances between k nearest neighbors in the dataset (k=30) (NH: neighbourhood). 
Neighbourhood statistics were calculated using a simulated trajectory dataset of 5000 cells. All plots show results from three sampling initializations for each proportion.

<!-- ```{r sup-fig-1, fig.pos="ht", fig.cap="(ref:sup-fig1)", echo=FALSE} -->
<!-- knitr::include_graphics("suppl_figs/suppl_fig1_letters.pdf", auto_pdf = TRUE) -->
<!-- ``` -->

<!-- (ref:sup-fig1) **Tolerance of Milo to uncorrected batch effects in simulated data** -->
<!-- (A-B) UMAP embeddings of simulated scRNA-seq data containing a batch effect, before batch correction (top row) and after correction with fastMNN (bottom row) (5000 cells). Cells are colored by simulated batch (A) and by presence of differential abundance between 2 simulated conditions (20% cells in condition ‘A’, 80% cells in condition ‘B’) (B). -->
<!-- (C) A graph representation of the results from Milo differential abundance testing. Neighbourhoods were tested for DA between conditions, with (~ batch + condition) or without (~ condition) accounting for the simulated batch. Nodes are neighbourhoods, coloured by their log fold change between conditions. Non-DA neighbourhoods (FDR 10%). Node sizes correspond to the number of cells in a neighbourhood. Graph edges depict the number of cells shared between adjacent neighbourhoods. -->
<!-- (D) Confusion matrices comparing the number of true and predicted DA neighbourhoods, with different batch effect correction (rows) and different testing design (columns). -->

<!-- ```{r sup-fig-2, fig.pos="ht", fig.cap="(ref:sup-fig2)", echo=FALSE} -->
<!-- knitr::include_graphics("suppl_figs/suppl_fig2_letters.pdf", auto_pdf = TRUE) -->
<!-- ``` -->

<!-- (ref:sup-fig2) **Tolerance of Milo to uncorrected batch effects in mouse gastrulation atlas** -->
<!-- (A-B) UMAP embedding of mouse gastrulation atlas before batch correction (top row) and after correction with fastMNN (bottom row). Cells are colored by sequencing batch (A) and developmental time point (B). -->
<!-- (C) Barplot depicting the percentage of DA neighbourhoods at FDR 10%, testing for different experimental covariates: DA between sequencing batches (~ Seq. batch), DA across developmental time points (~ Time points), DA across developmental time points accounting for the sequencing batch (~ Seq. batch + time points). The total number of DA neighbourhoods is shown in each bar. -->

```{r sup-fig-clustering, fig.pos="ht", fig.cap="(ref:sup-fig-clustering)", echo=FALSE}
knitr::include_graphics("suppl_figs/suppl_fig_clustering.pdf", auto_pdf = TRUE)
```

(ref:sup-fig-clustering) **Graph-clustering does not faithfully capture simulated groups and differentially abundant subpopulations in a simulated continuous trajectory.**
(A) A simulated linear trajectory of 2000 single-cells generated from 5 different groups, with cells assigned to either condition ‘A’ (left) or condition ‘B’ (right).
(B) A Walktrap clustering of the data in (A) using the same k-NN graph. Cells are coloured by Walktrap cluster identity.
(C) A Louvain clustering of the data in (A) using the same k-NN graph. Cells are coloured by the Louvain clustering identity.
(D-E) Heatmaps comparing the numbers of cells in each cluster with respect to the ground truth groups in (A). Each cell is coloured by the proportion of cells from the column groups (ground truth) that are assigned to the respective cluster.

```{r sup-fig-Kselection, fig.pos="ht", fig.cap="(ref:sup-fig-Kselection)", echo=FALSE}
knitr::include_graphics("suppl_figs/suppl_fig_Kselection.pdf", auto_pdf = TRUE)
```

(ref:sup-fig-Kselection) **Selection of $K$ parameter** 
(A) Illustrative examples of TPR-FDR trade-off for increasing values of $K$ used for k-NN graph building on simulated DA on 4 regions. Dotted lines highlight TPR=0.8 and FDR=0.1 thresholds. 
(B) The median number of cells per experimental sample is a function of the neighbourhood size snsdivided by the total number of samples (S). (C) Histogram of neighbourhood sizes for different choices of $K$. The red dotted line denotes the minimum neighbourhood size to obtain 5 cells per sample on average. 

```{r sup-fig-robustness, fig.pos="ht", fig.cap="(ref:sup-fig-robustness)", echo=FALSE}
knitr::include_graphics("suppl_figs/suppl_fig_robustness.pdf", auto_pdf = TRUE)
```

(ref:sup-fig-robustness) **Robustness of Milo DA testing to varying K**. Distributions of DA neighbourhoods across values of $K$ for the mouse ageing thymus (A) and human cirrhotic liver (B) data sets. Shown are the distributions of log fold-changes (y-axis) for DA (FDR 10%) neighbourhoods using different values of $K$ (x-axis) from 5-100, illustrating that DA testing is robust across a broad range of values of $K$.

```{r sup-fig-bm-threshold, fig.pos="ht", fig.cap="(ref:sup-fig-bm-threshold)", echo=FALSE}
knitr::include_graphics("suppl_figs/suppl_fig_bm_threshold.pdf", auto_pdf = TRUE)
```

(ref:sup-fig-bm-threshold) **Selection of probability threshold for DA benchmarking.** Mean True Positive Rate (left) and False Discovery Rate (right) for recovery of cells in simulated DA regions as a function of probability threshold $t$ picked to define true DA. The dashed line indicates $t = 0.6$, that was selected for benchmarking analyses. The mean is calculated over 81 simulations on 9 populations. Line shading indicates the standard deviation of the mean.

```{r sup-fig-bm-simulations, fig.pos="ht", fig.cap="(ref:sup-fig-bm-simulations)",  out.width = "500px", echo=FALSE}
knitr::include_graphics("suppl_figs/suppl_fig_bm_simulations.pdf", auto_pdf = TRUE)
```

(ref:sup-fig-bm-simulations) **Benchmarking DA methods on simulated data.** DA analysis performance on KNN graphs from simulated datasets of different topologies: (A) discrete clusters (900 cells, 3 populations);  (B) 1-D linear trajectory (2500 cells, 7 populations); (C) Branching trajectory (2500 cells, 10 populations).

```{r sup-fig-bm-batcheffects, fig.pos="ht", fig.cap="(ref:sup-fig-bm-batcheffects)", echo=FALSE}
knitr::include_graphics("suppl_figs/suppl_fig_bm_batcheffects.pdf", auto_pdf = TRUE)
```

(ref:sup-fig-bm-batcheffects) **Tolerance of DA analysis to batch effects.** (A) Comparison of performance of DA methods with no batch effect, with batch effects of increasing magnitude corrected with `fastMNN`, and uncorrected batch effects. (B) Comparison of Milo performance with (~ batch + condition) or without (~ condition) accounting for the simulated batch in the GLM. Points denote the Mean True Positive Rate (left) and False Discovery Rate (right) for recovery of cells in simulated DA regions with simulated batch effects of increasing magnitude. The mean is calculated over 27 simulations on 9 populations. Line shading indicates the standard deviation of the mean. Each panel represents a different DA effect size (Max C1 probability). 

```{r sup-fig-scalability, fig.pos="ht", fig.cap="(ref:sup-fig-scalability)",   out.width = "500px",echo=FALSE}
knitr::include_graphics("suppl_figs/suppl_fig_scalability.pdf", auto_pdf = TRUE)
```

(ref:sup-fig-scalability) **Memory usage across the Milo analysis workflow.**
Total memory usage across the steps of the Milo analysis workflow in 4 datasets containing different numbers of cells (Gastrulation: circles, Liver: triangles, Thymus: crosses, Simulation: squares). Grey points denote down-sampled datasets of the corresponding type. Coloured points denote the total number of cells for the respective dataset. Total memory usage (y-axis) is shown in megabytes (MB). (A) k-NN graph building, (B) neighbourhood sampling and construction, (C) within-neighbourhood distance calculation, (D) cell counting in neighbourhoods according to the input experimental design, (E) differential abundance testing, (F) total in memory R object size. A fixed value was used in all datasets for graph building and neighbourhood construction (k=30).

```{r sup-fig-thymus-nhoods, fig.pos="ht", fig.cap="(ref:sup-fig-thymus-nhoods)", echo=FALSE}
knitr::include_graphics("suppl_figs/suppl_fig_thymusNhoods.pdf", auto_pdf = TRUE)
```

(ref:sup-fig-thymus-nhoods) **Label transferred neighbourhood groups onto droplet scRNA-seq cells.** (A) Joint UMAP embedding for SMART-seq and droplet scRNA-seq datasets, points are coloured by label-transferred neighbourhood groups for the droplet scRNA-seq cells. (B) Proportions of label-transferred neighbourhood groups across mouse ages (n=3 replicates per age), corresponding to (A).


```{r sup-fig-liver-endo, fig.pos="ht", fig.cap="(ref:sup-fig-liver-endo)", echo=FALSE}
knitr::include_graphics("suppl_figs/suppl_fig_liver_endo.pdf", auto_pdf = TRUE)
```

(ref:sup-fig-liver-endo) **Downstream analysis of disease-specific endothelial subpopulations in liver cirrhosis**
(A) GO term enrichment analysis on marker genes of cirrhosis-enriched endothelia. (B) GO term enrichment analysis on marker genes of healthy-enriched endothelia. The top 20 significant terms are shown.

```{r sup-fig-liver-chol, fig.pos="ht", fig.cap="(ref:sup-fig-liver-chol)", echo=FALSE}
knitr::include_graphics("suppl_figs/suppl_fig_liver_chol.pdf", auto_pdf = TRUE)
```

(ref:sup-fig-liver-chol) **Downstream analysis of disease-specific cholangiocyte subpopulations in liver cirrhosis**
Downstream analysis of disease-specific cholangiocyte subpopulations in liver cirrhosis
(A-B) UMAP embedding (A) and graph representation (B) of neighbourhoods of 3369 cells from cholangiocyte lineage.
(C) Volcano plot for DGE test on cholangiocytes DA subpopulations: the x-axis shows the log-fold change between expression in cirrhotic and healthy cholangiocytes. The y-axis shows the adjusted p-value.
(D) GO term enrichment analysis on marker genes of cirrhosis-enriched cholangiocytes. The top 20 significant terms are shown.

<!-- ```{r sup-fig-6, fig.pos="ht", fig.cap="(ref:sup-fig6)", echo=FALSE} -->
<!-- knitr::include_graphics("suppl_figs/suppl_fig6.pdf", auto_pdf = TRUE) -->
<!-- ``` -->

<!-- (ref:sup-fig6) **Downstream analysis of disease-specific subpopulations in liver cirrhosis** -->
<!-- (A) GO term enrichment analysis on marker genes of cirrhosis-enriched endothelia. The top 20 significant terms are shown. -->
<!-- (B) GO term enrichment analysis on marker genes of healthy-enriched endothelia. The top 20 significant terms are shown. -->
<!-- (C-D) UMAP embedding (C) and graph representation (D) of neighbourhoods of 3369 cells from cholangiocytes lineage. -->
<!-- (E) Volcano plot for DGE test on cholangiocytes DA subpopulations: the x-axis shows the log-fold change between expression in cirrhotic and healthy cholangiocytes. The y-axis shows the adjusted p-value. -->
<!-- (F) GO term enrichment analysis on marker genes of cirrhosis-enriched cholangiocytes. The top 20 significant terms are shown. -->

<!-- ```{r sup-fig-6, fig.pos="ht", fig.cap="(ref:sup-fig6)", echo=FALSE} -->
<!-- knitr::include_graphics("suppl_figs/suppl_fig6.pdf", auto_pdf = TRUE) -->
<!-- ``` -->


<!-- \newpage -->

<!-- # Supplementary tables -->

<!-- | Method           | Key Parameters | Values             | Hypothesis testing                      | -->
<!-- |------------------|----------------|--------------------|-----------------------------------------| -->
<!-- | Milo             | K              | 10                 | Negative binomial GLM, 10% FDR          | -->
<!-- |                  | d              | 15                 | Negative binomial GLM, 10% FDR          | -->
<!-- | Cydar            | r              | 2                  | Negative binomial GLM, 10% FDR          | -->
<!-- | DAseq            | k.vector       | 5-500, steps of 50 | Logistic classifier prediction, top 10% | -->
<!-- | Louvain + edgeR  | K              | 10                 | Negative binomial GLM, 10% FDR          | -->
<!-- |                  | d              | 15                 |                                         | -->
<!-- | Walktrap + edgeR | K              | 10                 | Negative binomial GLM, 10% FDR          | -->
<!-- |                  | d              | 15                 |                                         | -->

<!-- Supplementary Table 1: **Method comparison parameter values.** -->

<!-- \newpage  -->

