---
title: "Supplementary information for \n**Differential cell-state abundance testing using k-NN graphs with _Milo_**"
author: 
  - Emma Dann, 
  - Neil C. Henderson,
  - Sarah A. Teichmann,
  - Michael D. Morgan,
  - John C. Marioni
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  bookdown::pdf_document2:
    keep_tex: yes
    toc: true
    toc_depth: 3
    number_sections: true
    includes:
      in_header:
        preamble.tex
bibliography: milo_supplement.bib
csl: cell-numeric.csl
link-citations: true
---

\renewcommand{\figurename}{Supplementary Figure}

\newpage 

# Supplementary notes 

## Description of workflow for _Milo_ analysis

Given a single-cell dataset of gene expression profiles of $M$ cells collected from $S$ experimental samples, _Milo_ aims to quantify systematic changes in abundance of cells between biological conditions, as compared to within-condition variability. Here we provide a step-by-step description of the workflow for differential abundance analysis.

### Preprocessing and dimensionality reduction

For preprocessing of scRNA-seq profiles we recommend following standard practices in single-cell analysis [@lueckenCurrentBestPractices2019; @amezquitaOrchestratingSinglecellAnalysis2020]: we normalize UMI counts by the total number of counts per cell, apply log-transformation and identify highly variable genes (HVGs). Then we project the $H \times M$ gene expression matrix, where $M$ is the number of cells and $H$ is the number of HVGs, to the first $d$ principal components (PCs). While downstream analysis is generally robust to the exact choice of the number of HVGs [@lueckenCurrentBestPractices2019], an optimal value for $d$ can be selected by detecting the "elbow" in the variance explained by PCs or using the "jackstraw" method [@chungStatisticalSignificanceVariables2015].

### Minimizing batch effects

Comparing biological conditions often requires acquiring single-cell data from multiple samples, that can be generated with different experimental conditions or protocols. This commonly introduces batch effects, which can have a substantial impact on the data composition and subsequently the topology of any k-NN graph computed across the single-cell data. Consequently, this will have an impact on the ability of _Milo_ to resolve genuine differential abundance of cells between experimental conditions of interest. In addition, other biological nuisance covariates could impact DA analysis i.e. biological factors that are not of interest for the analyst, such as donor of origin or sex of the donor. We recommend to mitigate the impact of technical or other nuisance covariates _before_ building the k-NN graph, by using one of the many _in silico_ integration tools designed for this task in single-cell datasets. 
<!-- We showed on an  that this step increases the performance of all DA testing methods (Suppl. Fig. <....>).  -->
Defining the best tool for this task is beyond the scope of this work, a large number of integration methods have been reviewed and benchmarked in [@lueckenBenchmarkingAtlaslevelData2020;@chazarra-gilFlexibleComparisonBatch2020;@tranBenchmarkBatcheffectCorrection2020]. However, users should consider the type of output produced by their integration method of choice, typically one of (A) a corrected feature space, (B) a joint embedding or (C) an integrated graph. The refined neighbourhood search procedure in _Milo_ relies on finding neighbors in reduced dimension space. Therefore using a methods that produces an integrated graph (e.g. BBKNN [@polanskiBBKNNFastBatch], Conos [@barkasJointAnalysisHeterogeneous2019]) could lead to suboptimal results in DA testing with _Milo_, because the refined neighbourhood search procedure would still be affected by the batch effect.

In addition, the effect of nuisance covariates should be modelled in the generalized linear model used for DA testing in _Milo_ to minimize the emergence of false positives in case of imperfect batch correction (see Section \@ref(testDA)) (Fig.2D).

We wish to emphasize that, when confounders are present, an appropriate experimental design is crucial to obtain reliable results from differential abundance analysis: if nuisance factors are 100% confounded with the biological condition used for differential abundance (e.g. if the samples from diseased and healthy donors are processed in separate sequencing batches), there is no way to disentangle the abundance differences that are truly driven by the biology of interest. We note that in a similar case applying a batch integration strategy before graph construction could lead to a loss of biological signal. 

<!-- To minimize the emergence of false positive and false negative results that might be introduced by such technical effects, we recommend the following procedure: (1) Batch effects should be mitigated with computational data integration. Defining the best tool for this task is beyond the scope of this work (a large number of integration methods have been reviewed and benchmarked in [@lueckenBenchmarkingAtlaslevelData2020;@chazarra-gilFlexibleComparisonBatch2020;@tranBenchmarkBatcheffectCorrection2020]). However, users should consider the type of output obtained by their integration method of choice, which can be a corrected feature space, a joint embedding or an integrated graph. Using a methods that produces a graph (e.g. BBKNN [@polanskiBBKNNFastBatch], Conos [@barkasJointAnalysisHeterogeneous2019]) will lead to suboptimal results in DA testing with _Milo_ as the refined neighbourhood search procedure will still be affected by the batch effect, as this relies on finding neighbors in PCA space.  -->

### Building the k-NN graph 

_Milo_ uses a k-NN graph computed based on similarities in gene expression space as a representation of the phenotypic manifold in which cells lie. While _Milo_ can be used on graphs built with different similarity kernels, here we compute the graph as follows: given the reduced dimension matrix $X_{PC}$ of dimensions $M \times d$, for each cell $j$, the Euclidean distances to its $K$ nearest neighbors in $X_{PC}$ are computed and stored in a $M \times M$ adjacency matrix $D$. Then, $D$ is made symmetrical, such that cells $j$ and $l$ are nearest neighbors (i.e. connected by an edge) if either $j$ is a nearest neighbor of $l$ or $l$ is a nearest neighbor of $j$. The k-NN graph is encoded by the undirected symmetric version of $\tilde{D}$ of $D$, where each cell has at least K nearest neighbors.

### Definition of cell neighbourhoods and index sampling algorithm

Next, we identify a set of representative cell neighbourhoods on the k-NN graph. We define the neighbourhood $n_i$ of cell $c_i$ as the group of cells that are connected to $c_i$ by an edge in the graph. We refer to $c_i$ with $i=1,2,...,N$ as the index cell of the neighbourhood, so that $N \leq M$.
<!-- Here, $i=1,2,...,N$ indexes the sampled neighbourhoods, while $j=1,2,...,M$ indexes the single-cells in the graph, so that $N \leq M$. -->
Formally, a cell $c_j$ belongs to neighbourhood $n_i$ if $\tilde{D}_{i,j} > 0$. 
<!-- We refer to $i$ as the index cell of the neighbourhood.  -->

In order to define neighbourhoods that span the whole k-NN graph, we sample index cells by using an algorithm previously adopted for waypoint sampling for trajectory inference [@gutTrajectoriesCellcycleProgression2015; @settyWishboneIdentifiesBifurcating2016].
Briefly, we start by randomly sampling $p \cdot M$ cells from the dataset, where $p \in [0,1]$  (we use $p = 0.1$ by default). 
Given the reduced dimension matrix used for graph construction $X_{PC}$, for each sampled cell $c_j$ we consider its $K$ nearest neighbors with PC profiles ${x_1, x_2, ... , x_k}$ and compute the mean position of the neighbors in PC space $\bar{x}$:
$$
\bar{x_j} = \frac{ \sum_k x_k }{K}
$$

Then, we search for the cell $c_i$ such that the Euclidean distance between $x_i$ and $\bar{x}$ is minimized. Because the algorithm might converge to the same index cell from multiple initial samplings, this procedure yields a set of $N \leq p \cdot M$ index cells that are used to define neighbourhoods. 

Having defined a set of $N$ neighbourhoods from the sampled index cells, we construct a count matrix of dimensions $N \times S$ which reports, for each sample, the number of cells that are present in each neighbourhood.

### Testing for differential abundance in neighbourhoods {#testDA}

To test for differential abundance between biological conditions, _Milo_ models the cell counts in neighbourhoods, estimating variability across biological replicates using a generalized linear model (GLM). We build upon the framework for differential abundance testing implemented by _Cydar_ [@lunTestingDifferentialAbundance2017]. In this section, we briefly describe the statistical model and adaptations to the k-NN graph setting. 

#### Quasi-likelihood negative binomial generalized linear models {-}

<!-- For neighbourhood $n$ with counts $y_{ns}$ for each sample $s$, the counts are modelled as a quasi-negative binomial distribution with mean $\mu_{ns}$ with a generalized linear model. This means that we allow for overdispersion of the counts data. -->

<!-- With a log-link GLM with G coefficients (which represent groups) the mean is -->
<!-- $$ -->
<!-- log\ \mu_{ns} = \sum_{g=1}^{G}x_{sg}\beta_{ng} + o_{ns} -->
<!-- $$ -->
<!-- Here $\beta_{ng}$ is the neighbourhood-specific value of coefficient $g$ i.e. the log-transformed average proportion of cells within a group. $x_{sg}$ specifies the group to which each sample belongs. $o_{ns}$ is the offset, representing the log-transformed total number of cells, that insures that differences in the numbers of cells between groups do not cause differences in $\beta_{ng}$. -->

We consider a neighbourhood $n$ with cell counts $y_{ns}$ for each experimental sample $s$. The counts are modelled by the negative binomial (NB) distribution, as it is supported over all non-negative integers and can accurately model both small and large cell counts. For such non-Normally distributed data we use generalized-linear models (GLMs) as an extension of classic linear models that can accomodate complex experimental designs. We therefore assume that
$$
y_{ns} \sim NB(\mu_{ns},\phi_{n}),
$$
where $\mu_{ns}$ is the mean number of cells from sample $s$ in neighbourhood $n$ and $\phi_{n}$ is the NB dispersion parameter.

The expected count value $\mu_{ns}$ is given by 
$$
\mu_{ns} = \lambda_{ns}M_s
$$

where $\lambda_{ns}$ is the proportion of cells belonging to experimental sample $s$ in $n$ and $M_s$ is the total number of cells of $s$. In practice, $\lambda_{ns}$ represents the biological variability that can be affected by treatment condition, age or any biological covariate of interest. 

We use a log-linear model to model the influence of a biological condition on the expected counts in the neighbourhood:
$$
log\ \mu_{ns} = \sum_{g=1}^{G}x_{sg}\beta_{ng} + log\ M_s
$$
Here, for each possible value $g$ taken by the biological condition of interest, $x_{sg}$ is the binary vector indicating the condition value applied to sample $s$. $\beta_{ng}$ is the regression coefficient by which the covariate effects are mediated for neighbourhood $n$, that represents the log fold-change between number of cells in condition $g$ and all other conditions. If the biological condition of interest is ordinal (such as age or disease-severity) $\beta_{ng}$ is interpreted as the per-unit linear change in neighbourhood abundance.

Estimation of $\beta_{ng}$ for each $n$ and $g$ is performed by fitting the GLM to the count data for each neighbourhood, i.e. by estimating the dispersion $\phi_{n}$ that models the variability of cell counts for replicate samples for each neighbourhood. Dispersion estimation is performed using the quasi-likelihood method in `edgeR`[@robinsonEdgeRBioconductorPackage2010a], where the dispersion is modelled from the GLM deviance and thereby stabilized with empirical Bayes shrinkage, to stabilize the estimates in the presence of limited replication.

#### Adaptation of Spatial FDR to neighbourhoods {-}

WeTo account for the non-independence of spatially overlapping neighbourhoods by applying a weighted version of the Benjamini-Hochberg (BH) method, where P values are weighted by the reciprocal of the neighbourhood connectivity, we build upon  as an adaptation to graphs of a previously described strategy to control the spatial False Discovery Rate (FDR)  [6]. 

To control for multiple testing, we need to account for the overlap between neighbourhoods, that makes the differential abundance tests non-independent. We apply a weighted version of the Benjamini-Hochberg (BH) method, where P values are weighted by the reciprocal of the neighbourhood connectivity, as an adaptation to graphs of the Spatial FDR method introduced by $Cydar$ [@lunTestingDifferentialAbundance2017].
Formally, to control for FDR at a selected threshold $\alpha$ we reject null hypothesis $i$ where the associated p-value is less than the threshold:

$$
\max_i{p_{(i)}: p_{(i)}\le \alpha\frac{\sum_{l=1}^{i}w_{(l)}}{\sum_{l=1}^{n}w_{(l)}}}
$$
Where the weight $w_{(i)}$ is the reciprocal of the neighbourhood connectivity $c_i$. As a measure of neighbourhood connectivity, we use the Euclidean distance between the neighbourhood index cell $c_i$ and its kth nearest neighbour in PC space.

## Guidelines for choice of parameters

.... 

<!-- ## Accounting for batch effects -->

<!-- To demonstrate the use of including both steps in DA testing analysis, we simulated data for a continuous trajectory with two technical batches (Suppl.Fig.2A) and simulated a subpopulation of differential abundance between two conditions (Suppl.Fig.2B). We tested for DA between conditions on the k-NN graph generated before and after batch correction using the Mutual Nearest Neighbors (MNN) method [@haghverdiBatchEffectsSinglecell2018]. In both the corrected and uncorrected data, we examined the effect of accounting for the batch in the experimental design of the GLM (`design = ~ batch + condition`). We find that, while alone MNN correction increases the true positive rate, accounting for the nuisance covariate in the testing design significantly increases accuracy and power both in the uncorrected and corrected graph (Suppl. Fig. 2C-D). The best DA prediction performance was achieved combining batch correction and an explicit experimental design adjusting for such nuisance covariates (Suppl. Fig. 2D). -->

<!-- To demonstrate how these results apply to a real dataset with a complex experimental design, we used a single-cell atlas of mouse gastrulation @pijuan-salaSinglecellMolecularMap2019. These data consist of 116312 cells collected from 411 embryos in 36 samples, across 10 developmental stages from E6.5 to E8.5 (Suppl. Fig. 3B). The samples were sequenced in 3 batches (Suppl. Fig. 3A). We used _Milo_ to test for DA across developmental time points, on the k-NN graph built before and after batch correction with MNN. We find that including the sequencing batch in the DA testing design matrix removes false positives in the uncorrected graph, while increasing the number of detected DA neighbourhoods in the MNN corrected graph (Suppl. Fig.3C). -->

<!-- Taken together these results illustrate that while _Milo_ can be used to robustly account for batch effects during DA testing, optimal results are achieved after the removal of major batch effects that are not readily accounted for by inclusion as a covariate in a linear model. -->


# References


